

<!DOCTYPE html>
<html class="writer-html5" lang="en" data-content_root="../../">
<head>
  <meta charset="utf-8" />
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>neurosnap.protein &mdash; Neurosnap 2025.01.08 documentation</title>
      <link rel="stylesheet" type="text/css" href="../../_static/pygments.css?v=b86133f3" />
      <link rel="stylesheet" type="text/css" href="../../_static/css/theme.css?v=e59714d7" />

  
    <link rel="shortcut icon" href="../../_static/favicon.ico"/>
      <script src="../../_static/jquery.js?v=5d32c60e"></script>
      <script src="../../_static/_sphinx_javascript_frameworks_compat.js?v=2cd50e6c"></script>
      <script src="../../_static/documentation_options.js?v=ad7aade1"></script>
      <script src="../../_static/doctools.js?v=9bcbadda"></script>
      <script src="../../_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="../../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../../genindex.html" />
    <link rel="search" title="Search" href="../../search.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="../../index.html" class="icon icon-home">
            Neurosnap
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Documentation</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../modules.html">neurosnap</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../index.html">Neurosnap</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../../index.html" class="icon icon-home" aria-label="Home"></a></li>
          <li class="breadcrumb-item"><a href="../index.html">Module code</a></li>
      <li class="breadcrumb-item active">neurosnap.protein</li>
      <li class="wy-breadcrumbs-aside">
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <h1>Source code for neurosnap.protein</h1><div class="highlight"><pre>
<span></span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">Provides functions and classes related to processing protein data as well as</span>
<span class="sd">a feature rich wrapper around protein structures using BioPython.</span>
<span class="sd">&quot;&quot;&quot;</span>

<span class="kn">import</span><span class="w"> </span><span class="nn">io</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">json</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">os</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">shutil</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">tempfile</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">time</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">xml.etree.ElementTree</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">ET</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">typing</span><span class="w"> </span><span class="kn">import</span> <span class="n">List</span><span class="p">,</span> <span class="n">Optional</span><span class="p">,</span> <span class="n">Tuple</span><span class="p">,</span> <span class="n">Union</span>

<span class="kn">import</span><span class="w"> </span><span class="nn">matplotlib</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">matplotlib.animation</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">animation</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">matplotlib.patheffects</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">matplotlib.pyplot</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">plt</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">numpy</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">np</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">pandas</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">pd</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">requests</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">Bio.PDB</span><span class="w"> </span><span class="kn">import</span> <span class="n">PDBIO</span><span class="p">,</span> <span class="n">SASA</span><span class="p">,</span> <span class="n">PDBParser</span><span class="p">,</span> <span class="n">PPBuilder</span><span class="p">,</span> <span class="n">MMCIFParser</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">Bio.PDB.mmcifio</span><span class="w"> </span><span class="kn">import</span> <span class="n">MMCIFIO</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">Bio.PDB.Polypeptide</span><span class="w"> </span><span class="kn">import</span> <span class="n">is_aa</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">Bio.PDB.Superimposer</span><span class="w"> </span><span class="kn">import</span> <span class="n">Superimposer</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">matplotlib</span><span class="w"> </span><span class="kn">import</span> <span class="n">collections</span> <span class="k">as</span> <span class="n">mcoll</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">rdkit</span><span class="w"> </span><span class="kn">import</span> <span class="n">Chem</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">requests_toolbelt.multipart.encoder</span><span class="w"> </span><span class="kn">import</span> <span class="n">MultipartEncoder</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">scipy.special</span><span class="w"> </span><span class="kn">import</span> <span class="n">expit</span> <span class="k">as</span> <span class="n">sigmoid</span>

<span class="kn">import</span><span class="w"> </span><span class="nn">neurosnap.algos.lDDT</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">lDDT</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">neurosnap.api</span><span class="w"> </span><span class="kn">import</span> <span class="n">USER_AGENT</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">neurosnap.log</span><span class="w"> </span><span class="kn">import</span> <span class="n">logger</span>

<span class="c1">### CONSTANTS ###</span>
<span class="c1"># Names of atoms that are part of a protein&#39;s backbone structure</span>
<span class="n">BACKBONE_ATOMS</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;N&quot;</span><span class="p">,</span> <span class="s2">&quot;CA&quot;</span><span class="p">,</span> <span class="s2">&quot;C&quot;</span><span class="p">}</span>
<span class="c1"># Codes for both RNA and DNA</span>
<span class="n">STANDARD_NUCLEOTIDES</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;A&quot;</span><span class="p">,</span> <span class="s2">&quot;T&quot;</span><span class="p">,</span> <span class="s2">&quot;C&quot;</span><span class="p">,</span> <span class="s2">&quot;G&quot;</span><span class="p">,</span> <span class="s2">&quot;U&quot;</span><span class="p">,</span> <span class="s2">&quot;DA&quot;</span><span class="p">,</span> <span class="s2">&quot;DT&quot;</span><span class="p">,</span> <span class="s2">&quot;DC&quot;</span><span class="p">,</span> <span class="s2">&quot;DG&quot;</span><span class="p">,</span> <span class="s2">&quot;DU&quot;</span><span class="p">}</span>
<span class="c1">## Standard amino acids excluding the unknown character (&quot;X&quot;)</span>
<span class="c1">## Currently excludes</span>
<span class="c1"># O | pyl | pyrrolysine</span>
<span class="c1"># U | sec | selenocysteine</span>
<span class="c1"># B | asx | asparagine/aspartic acid</span>
<span class="c1"># Z | glx | glutamine/glutamic acid</span>
<span class="c1"># J | xle | leucine/isoleucine</span>
<span class="c1"># X | UNK | unknown codon</span>
<span class="c1"># * | TRM | termination codon</span>
<span class="n">STANDARD_AAs</span> <span class="o">=</span> <span class="s2">&quot;ACDEFGHIKLMNPQRSTVWY&quot;</span>
<span class="n">STANDARD_AAs_ABR</span> <span class="o">=</span> <span class="p">{</span>
  <span class="s2">&quot;ALA&quot;</span><span class="p">,</span>
  <span class="s2">&quot;ARG&quot;</span><span class="p">,</span>
  <span class="s2">&quot;ASN&quot;</span><span class="p">,</span>
  <span class="s2">&quot;ASP&quot;</span><span class="p">,</span>
  <span class="s2">&quot;CYS&quot;</span><span class="p">,</span>
  <span class="s2">&quot;GLN&quot;</span><span class="p">,</span>
  <span class="s2">&quot;GLU&quot;</span><span class="p">,</span>
  <span class="s2">&quot;GLY&quot;</span><span class="p">,</span>
  <span class="s2">&quot;HIS&quot;</span><span class="p">,</span>
  <span class="s2">&quot;ILE&quot;</span><span class="p">,</span>
  <span class="s2">&quot;LEU&quot;</span><span class="p">,</span>
  <span class="s2">&quot;LYS&quot;</span><span class="p">,</span>
  <span class="s2">&quot;MET&quot;</span><span class="p">,</span>
  <span class="s2">&quot;PHE&quot;</span><span class="p">,</span>
  <span class="s2">&quot;PRO&quot;</span><span class="p">,</span>
  <span class="s2">&quot;SER&quot;</span><span class="p">,</span>
  <span class="s2">&quot;THR&quot;</span><span class="p">,</span>
  <span class="s2">&quot;TRP&quot;</span><span class="p">,</span>
  <span class="s2">&quot;TYR&quot;</span><span class="p">,</span>
  <span class="s2">&quot;VAL&quot;</span><span class="p">,</span>
<span class="p">}</span>
<span class="c1"># List of hydrophobic residues</span>
<span class="n">HYDROPHOBIC_RESIDUES</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;ALA&quot;</span><span class="p">,</span> <span class="s2">&quot;VAL&quot;</span><span class="p">,</span> <span class="s2">&quot;LEU&quot;</span><span class="p">,</span> <span class="s2">&quot;ILE&quot;</span><span class="p">,</span> <span class="s2">&quot;MET&quot;</span><span class="p">,</span> <span class="s2">&quot;PHE&quot;</span><span class="p">,</span> <span class="s2">&quot;TRP&quot;</span><span class="p">,</span> <span class="s2">&quot;PRO&quot;</span><span class="p">}</span>

<span class="c1">## Full amino acids table</span>
<span class="n">AAs_FULL_TABLE</span> <span class="o">=</span> <span class="p">[</span>
  <span class="p">[</span><span class="s2">&quot;A&quot;</span><span class="p">,</span> <span class="s2">&quot;ALA&quot;</span><span class="p">,</span> <span class="s2">&quot;ALANINE&quot;</span><span class="p">],</span>
  <span class="p">[</span><span class="s2">&quot;R&quot;</span><span class="p">,</span> <span class="s2">&quot;ARG&quot;</span><span class="p">,</span> <span class="s2">&quot;ARGININE&quot;</span><span class="p">],</span>
  <span class="p">[</span><span class="s2">&quot;N&quot;</span><span class="p">,</span> <span class="s2">&quot;ASN&quot;</span><span class="p">,</span> <span class="s2">&quot;ASPARAGINE&quot;</span><span class="p">],</span>
  <span class="p">[</span><span class="s2">&quot;D&quot;</span><span class="p">,</span> <span class="s2">&quot;ASP&quot;</span><span class="p">,</span> <span class="s2">&quot;ASPARTIC ACID&quot;</span><span class="p">],</span>
  <span class="p">[</span><span class="s2">&quot;C&quot;</span><span class="p">,</span> <span class="s2">&quot;CYS&quot;</span><span class="p">,</span> <span class="s2">&quot;CYSTEINE&quot;</span><span class="p">],</span>
  <span class="p">[</span><span class="s2">&quot;Q&quot;</span><span class="p">,</span> <span class="s2">&quot;GLN&quot;</span><span class="p">,</span> <span class="s2">&quot;GLUTAMINE&quot;</span><span class="p">],</span>
  <span class="p">[</span><span class="s2">&quot;E&quot;</span><span class="p">,</span> <span class="s2">&quot;GLU&quot;</span><span class="p">,</span> <span class="s2">&quot;GLUTAMIC ACID&quot;</span><span class="p">],</span>
  <span class="p">[</span><span class="s2">&quot;G&quot;</span><span class="p">,</span> <span class="s2">&quot;GLY&quot;</span><span class="p">,</span> <span class="s2">&quot;GLYCINE&quot;</span><span class="p">],</span>
  <span class="p">[</span><span class="s2">&quot;H&quot;</span><span class="p">,</span> <span class="s2">&quot;HIS&quot;</span><span class="p">,</span> <span class="s2">&quot;HISTIDINE&quot;</span><span class="p">],</span>
  <span class="p">[</span><span class="s2">&quot;I&quot;</span><span class="p">,</span> <span class="s2">&quot;ILE&quot;</span><span class="p">,</span> <span class="s2">&quot;ISOLEUCINE&quot;</span><span class="p">],</span>
  <span class="p">[</span><span class="s2">&quot;L&quot;</span><span class="p">,</span> <span class="s2">&quot;LEU&quot;</span><span class="p">,</span> <span class="s2">&quot;LEUCINE&quot;</span><span class="p">],</span>
  <span class="p">[</span><span class="s2">&quot;K&quot;</span><span class="p">,</span> <span class="s2">&quot;LYS&quot;</span><span class="p">,</span> <span class="s2">&quot;LYSINE&quot;</span><span class="p">],</span>
  <span class="p">[</span><span class="s2">&quot;M&quot;</span><span class="p">,</span> <span class="s2">&quot;MET&quot;</span><span class="p">,</span> <span class="s2">&quot;METHIONINE&quot;</span><span class="p">],</span>
  <span class="p">[</span><span class="s2">&quot;F&quot;</span><span class="p">,</span> <span class="s2">&quot;PHE&quot;</span><span class="p">,</span> <span class="s2">&quot;PHENYLALANINE&quot;</span><span class="p">],</span>
  <span class="p">[</span><span class="s2">&quot;P&quot;</span><span class="p">,</span> <span class="s2">&quot;PRO&quot;</span><span class="p">,</span> <span class="s2">&quot;PROLINE&quot;</span><span class="p">],</span>
  <span class="p">[</span><span class="s2">&quot;S&quot;</span><span class="p">,</span> <span class="s2">&quot;SER&quot;</span><span class="p">,</span> <span class="s2">&quot;SERINE&quot;</span><span class="p">],</span>
  <span class="p">[</span><span class="s2">&quot;T&quot;</span><span class="p">,</span> <span class="s2">&quot;THR&quot;</span><span class="p">,</span> <span class="s2">&quot;THREONINE&quot;</span><span class="p">],</span>
  <span class="p">[</span><span class="s2">&quot;W&quot;</span><span class="p">,</span> <span class="s2">&quot;TRP&quot;</span><span class="p">,</span> <span class="s2">&quot;TRYPTOPHAN&quot;</span><span class="p">],</span>
  <span class="p">[</span><span class="s2">&quot;Y&quot;</span><span class="p">,</span> <span class="s2">&quot;TYR&quot;</span><span class="p">,</span> <span class="s2">&quot;TYROSINE&quot;</span><span class="p">],</span>
  <span class="p">[</span><span class="s2">&quot;V&quot;</span><span class="p">,</span> <span class="s2">&quot;VAL&quot;</span><span class="p">,</span> <span class="s2">&quot;VALINE&quot;</span><span class="p">],</span>
  <span class="p">[</span><span class="s2">&quot;O&quot;</span><span class="p">,</span> <span class="s2">&quot;PYL&quot;</span><span class="p">,</span> <span class="s2">&quot;PYRROLYSINE&quot;</span><span class="p">],</span>
  <span class="p">[</span><span class="s2">&quot;U&quot;</span><span class="p">,</span> <span class="s2">&quot;SEC&quot;</span><span class="p">,</span> <span class="s2">&quot;SELENOCYSTEINE&quot;</span><span class="p">],</span>
  <span class="p">[</span><span class="s2">&quot;B&quot;</span><span class="p">,</span> <span class="s2">&quot;ASX&quot;</span><span class="p">,</span> <span class="s2">&quot;ASPARAGINE/ASPARTIC ACID&quot;</span><span class="p">],</span>
  <span class="p">[</span><span class="s2">&quot;Z&quot;</span><span class="p">,</span> <span class="s2">&quot;GLX&quot;</span><span class="p">,</span> <span class="s2">&quot;GLUTAMINE/GLUTAMIC ACID&quot;</span><span class="p">],</span>
  <span class="p">[</span><span class="s2">&quot;J&quot;</span><span class="p">,</span> <span class="s2">&quot;XLE&quot;</span><span class="p">,</span> <span class="s2">&quot;LEUCINE/ISOLEUCINE&quot;</span><span class="p">],</span>
  <span class="p">[</span><span class="s2">&quot;X&quot;</span><span class="p">,</span> <span class="s2">&quot;UNK&quot;</span><span class="p">,</span> <span class="s2">&quot;UNKNOWN CODON&quot;</span><span class="p">],</span>
<span class="p">]</span>
<span class="n">AA_CODE_TO_ABR</span> <span class="o">=</span> <span class="p">{}</span>
<span class="n">AA_CODE_TO_NAME</span> <span class="o">=</span> <span class="p">{}</span>
<span class="n">AA_ABR_TO_CODE</span> <span class="o">=</span> <span class="p">{}</span>
<span class="n">AA_ABR_TO_NAME</span> <span class="o">=</span> <span class="p">{}</span>
<span class="n">AA_NAME_TO_CODE</span> <span class="o">=</span> <span class="p">{}</span>
<span class="n">AA_NAME_TO_ABR</span> <span class="o">=</span> <span class="p">{}</span>
<span class="k">for</span> <span class="n">code</span><span class="p">,</span> <span class="n">abr</span><span class="p">,</span> <span class="n">name</span> <span class="ow">in</span> <span class="n">AAs_FULL_TABLE</span><span class="p">:</span>
  <span class="n">AA_CODE_TO_ABR</span><span class="p">[</span><span class="n">code</span><span class="p">]</span> <span class="o">=</span> <span class="n">abr</span>
  <span class="n">AA_CODE_TO_NAME</span><span class="p">[</span><span class="n">code</span><span class="p">]</span> <span class="o">=</span> <span class="n">name</span>
  <span class="n">AA_ABR_TO_CODE</span><span class="p">[</span><span class="n">abr</span><span class="p">]</span> <span class="o">=</span> <span class="n">code</span>
  <span class="n">AA_ABR_TO_NAME</span><span class="p">[</span><span class="n">abr</span><span class="p">]</span> <span class="o">=</span> <span class="n">name</span>
  <span class="n">AA_NAME_TO_ABR</span><span class="p">[</span><span class="n">name</span><span class="p">]</span> <span class="o">=</span> <span class="n">abr</span>
  <span class="n">AA_NAME_TO_CODE</span><span class="p">[</span><span class="n">name</span><span class="p">]</span> <span class="o">=</span> <span class="n">code</span>


<span class="c1">### CLASSES ###</span>
<div class="viewcode-block" id="Protein">
<a class="viewcode-back" href="../../neurosnap.html#neurosnap.protein.Protein">[docs]</a>
<span class="k">class</span><span class="w"> </span><span class="nc">Protein</span><span class="p">:</span>
<div class="viewcode-block" id="Protein.__init__">
<a class="viewcode-back" href="../../neurosnap.html#neurosnap.protein.Protein.__init__">[docs]</a>
  <span class="k">def</span><span class="w"> </span><span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">pdb</span><span class="p">:</span> <span class="n">Union</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">io</span><span class="o">.</span><span class="n">IOBase</span><span class="p">],</span> <span class="nb">format</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;auto&quot;</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Class that wraps around a protein structure.</span>

<span class="sd">    Utilizes the biopython protein structure under the hood.</span>
<span class="sd">    Atoms that are not part of a chain will automatically be</span>
<span class="sd">    added to a new chain that does not overlap with any</span>
<span class="sd">    existing chains.</span>
<span class="sd">I</span>
<span class="sd">    Parameters:</span>
<span class="sd">      pdb: Can be either a file handle, PDB or mmCIF filepath, PDB ID, or UniProt ID</span>
<span class="sd">      format: File format of the input (&quot;pdb&quot;, &quot;mmcif&quot;, or &quot;auto&quot; to infer format from extension)</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">title</span> <span class="o">=</span> <span class="s2">&quot;Untitled Protein&quot;</span>
    <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">pdb</span><span class="p">,</span> <span class="n">io</span><span class="o">.</span><span class="n">IOBase</span><span class="p">):</span>
      <span class="k">pass</span>
    <span class="k">elif</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">pdb</span><span class="p">,</span> <span class="nb">str</span><span class="p">):</span>
      <span class="k">if</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">exists</span><span class="p">(</span><span class="n">pdb</span><span class="p">):</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">title</span> <span class="o">=</span> <span class="n">pdb</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;/&quot;</span><span class="p">)[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>
      <span class="k">else</span><span class="p">:</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">title</span> <span class="o">=</span> <span class="n">pdb</span><span class="o">.</span><span class="n">upper</span><span class="p">()</span>
        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">pdb</span><span class="p">)</span> <span class="o">==</span> <span class="mi">4</span><span class="p">:</span>  <span class="c1"># check if a valid PDB ID</span>
          <span class="n">r</span> <span class="o">=</span> <span class="n">requests</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;https://files.rcsb.org/download/</span><span class="si">{</span><span class="n">pdb</span><span class="o">.</span><span class="n">upper</span><span class="p">()</span><span class="si">}</span><span class="s2">.pdb&quot;</span><span class="p">)</span>
          <span class="n">r</span><span class="o">.</span><span class="n">raise_for_status</span><span class="p">()</span>
          <span class="k">with</span> <span class="n">tempfile</span><span class="o">.</span><span class="n">NamedTemporaryFile</span><span class="p">(</span><span class="n">delete</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">suffix</span><span class="o">=</span><span class="s2">&quot;.pdb&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">temp_file</span><span class="p">:</span>
            <span class="n">temp_file</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">r</span><span class="o">.</span><span class="n">content</span><span class="p">)</span>
            <span class="n">pdb</span> <span class="o">=</span> <span class="n">temp_file</span><span class="o">.</span><span class="n">name</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Found matching structure in RCSB PDB, downloading and using that.&quot;</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>  <span class="c1"># check if provided a uniprot ID and fetch from AF2</span>
          <span class="n">r</span> <span class="o">=</span> <span class="n">requests</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;https://alphafold.ebi.ac.uk/api/prediction/</span><span class="si">{</span><span class="n">pdb</span><span class="o">.</span><span class="n">upper</span><span class="p">()</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
          <span class="k">if</span> <span class="n">r</span><span class="o">.</span><span class="n">status_code</span> <span class="o">!=</span> <span class="mi">200</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;Invalid input type provided. Can be either a file handle, PDB filepath, PDB ID, or UniProt ID&quot;</span><span class="p">)</span>
          <span class="n">r</span> <span class="o">=</span> <span class="n">requests</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">r</span><span class="o">.</span><span class="n">json</span><span class="p">()[</span><span class="mi">0</span><span class="p">][</span><span class="s2">&quot;pdbUrl&quot;</span><span class="p">])</span>
          <span class="k">if</span> <span class="n">r</span><span class="o">.</span><span class="n">status_code</span> <span class="o">!=</span> <span class="mi">200</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;Invalid input type provided. Can be either a file handle, PDB filepath, PDB ID, or UniProt ID&quot;</span><span class="p">)</span>
          <span class="k">with</span> <span class="n">tempfile</span><span class="o">.</span><span class="n">NamedTemporaryFile</span><span class="p">(</span><span class="n">delete</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">suffix</span><span class="o">=</span><span class="s2">&quot;.pdb&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">temp_file</span><span class="p">:</span>
            <span class="n">temp_file</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">r</span><span class="o">.</span><span class="n">content</span><span class="p">)</span>
            <span class="n">pdb</span> <span class="o">=</span> <span class="n">temp_file</span><span class="o">.</span><span class="n">name</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Found matching structure in AF-DB, downloading and using that.&quot;</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
      <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;Invalid input type provided. Can be either a file handle, PDB filepath, PDB ID, or UniProt ID&quot;</span><span class="p">)</span>

    <span class="c1"># Infer format if set to auto</span>
    <span class="k">if</span> <span class="nb">format</span> <span class="o">==</span> <span class="s2">&quot;auto&quot;</span><span class="p">:</span>
      <span class="k">if</span> <span class="nb">str</span><span class="p">(</span><span class="n">pdb</span><span class="p">)</span><span class="o">.</span><span class="n">lower</span><span class="p">()</span><span class="o">.</span><span class="n">endswith</span><span class="p">(</span><span class="s2">&quot;.pdb&quot;</span><span class="p">):</span>
        <span class="nb">format</span> <span class="o">=</span> <span class="s2">&quot;pdb&quot;</span>
      <span class="k">elif</span> <span class="nb">str</span><span class="p">(</span><span class="n">pdb</span><span class="p">)</span><span class="o">.</span><span class="n">lower</span><span class="p">()</span><span class="o">.</span><span class="n">endswith</span><span class="p">(</span><span class="s2">&quot;.cif&quot;</span><span class="p">)</span> <span class="ow">or</span> <span class="nb">str</span><span class="p">(</span><span class="n">pdb</span><span class="p">)</span><span class="o">.</span><span class="n">lower</span><span class="p">()</span><span class="o">.</span><span class="n">endswith</span><span class="p">(</span><span class="s2">&quot;.mmcif&quot;</span><span class="p">):</span>
        <span class="nb">format</span> <span class="o">=</span> <span class="s2">&quot;mmcif&quot;</span>
      <span class="k">else</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;Failed to infer format. Please specify format explicitly as &#39;pdb&#39; or &#39;mmcif&#39;.&quot;</span><span class="p">)</span>

    <span class="c1"># load structure</span>
    <span class="k">if</span> <span class="nb">format</span> <span class="o">==</span> <span class="s2">&quot;pdb&quot;</span><span class="p">:</span>
      <span class="n">parser</span> <span class="o">=</span> <span class="n">PDBParser</span><span class="p">()</span>
    <span class="k">elif</span> <span class="nb">format</span> <span class="o">==</span> <span class="s2">&quot;mmcif&quot;</span><span class="p">:</span>
      <span class="n">parser</span> <span class="o">=</span> <span class="n">MMCIFParser</span><span class="p">()</span>
    <span class="k">else</span><span class="p">:</span>
      <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;Invalid format specified. Supported formats are &#39;pdb&#39; or &#39;mmcif&#39;.&quot;</span><span class="p">)</span>

    <span class="bp">self</span><span class="o">.</span><span class="n">structure</span> <span class="o">=</span> <span class="n">parser</span><span class="o">.</span><span class="n">get_structure</span><span class="p">(</span><span class="s2">&quot;structure&quot;</span><span class="p">,</span> <span class="n">pdb</span><span class="p">)</span>
    <span class="k">assert</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">structure</span><span class="p">),</span> <span class="s2">&quot;No models found. Structure appears to be empty.&quot;</span>

    <span class="c1"># Adds any missing chains to the structure.</span>
    <span class="c1"># Ensures new chain IDs do not overlap with existing ones.</span>
    <span class="n">existing_chains</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">chains</span><span class="p">()</span>
    <span class="n">alphabet</span> <span class="o">=</span> <span class="s2">&quot;ABCDEFGHIJKLMNOPQRSTUVWXYZ&quot;</span>
    <span class="n">numeric_suffix</span> <span class="o">=</span> <span class="mi">1</span>
    <span class="n">new_chain_id</span> <span class="o">=</span> <span class="kc">None</span>

    <span class="c1"># Find an available chain ID that does not overlap with existing ones</span>
    <span class="k">for</span> <span class="n">char</span> <span class="ow">in</span> <span class="n">alphabet</span><span class="p">:</span>
      <span class="k">if</span> <span class="n">char</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">existing_chains</span><span class="p">:</span>
        <span class="n">new_chain_id</span> <span class="o">=</span> <span class="n">char</span>
        <span class="k">break</span>
    <span class="k">if</span> <span class="n">new_chain_id</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
      <span class="c1"># If all single-letter chain IDs are used, append a numeric suffix</span>
      <span class="k">while</span> <span class="n">new_chain_id</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">char</span> <span class="ow">in</span> <span class="n">alphabet</span><span class="p">:</span>
          <span class="n">candidate_id</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">char</span><span class="si">}{</span><span class="n">numeric_suffix</span><span class="si">}</span><span class="s2">&quot;</span>
          <span class="k">if</span> <span class="n">candidate_id</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">existing_chains</span><span class="p">:</span>
            <span class="n">new_chain_id</span> <span class="o">=</span> <span class="n">candidate_id</span>
            <span class="k">break</span>
        <span class="n">numeric_suffix</span> <span class="o">+=</span> <span class="mi">1</span>

    <span class="k">if</span> <span class="n">new_chain_id</span><span class="p">:</span>
      <span class="c1"># rename chain</span>
      <span class="k">for</span> <span class="n">model</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">structure</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">chain</span> <span class="ow">in</span> <span class="n">model</span><span class="p">:</span>
          <span class="k">if</span> <span class="n">chain</span><span class="o">.</span><span class="n">id</span> <span class="ow">is</span> <span class="kc">None</span> <span class="ow">or</span> <span class="n">chain</span><span class="o">.</span><span class="n">id</span> <span class="o">==</span> <span class="s2">&quot; &quot;</span><span class="p">:</span>
            <span class="n">chain</span><span class="o">.</span><span class="n">id</span> <span class="o">=</span> <span class="n">new_chain_id</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Assigned ID &#39;</span><span class="si">{</span><span class="n">new_chain_id</span><span class="si">}</span><span class="s2">&#39; to chain with missing ID in model &#39;</span><span class="si">{</span><span class="n">model</span><span class="o">.</span><span class="n">id</span><span class="si">}</span><span class="s2">&#39;.&quot;</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
      <span class="n">logger</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;No chain IDs available. Could not rename chain with missing ID in model &#39;</span><span class="si">{</span><span class="n">model</span><span class="o">.</span><span class="n">id</span><span class="si">}</span><span class="s2">&#39;.&quot;</span><span class="p">)</span>

    <span class="c1"># generate the pandas dataframe similar to that of biopandas</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">generate_df</span><span class="p">()</span></div>


  <span class="k">def</span><span class="w"> </span><span class="fm">__repr__</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
    <span class="k">return</span> <span class="sa">f</span><span class="s2">&quot;&lt;Neurosnap Protein: Title=</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">title</span><span class="si">}</span><span class="s2"> Models=</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">models</span><span class="p">()</span><span class="si">}</span><span class="s2">, Chains=[</span><span class="si">{</span><span class="s1">&#39;, &#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">chains</span><span class="p">())</span><span class="si">}</span><span class="s2">], Atoms=</span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">df</span><span class="p">)</span><span class="si">}</span><span class="s2">&gt;&quot;</span>

<div class="viewcode-block" id="Protein.__call__">
<a class="viewcode-back" href="../../neurosnap.html#neurosnap.protein.Protein.__call__">[docs]</a>
  <span class="k">def</span><span class="w"> </span><span class="fm">__call__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">model</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">int</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">chain</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">int</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">res_type</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">int</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Returns a selection of a copy of the internal dataframe</span>
<span class="sd">    that matches the provided query. If no queries are</span>
<span class="sd">    provided, will return a copy of the internal dataframe.</span>

<span class="sd">    Parameters:</span>
<span class="sd">      model: If provided, returned atoms must match this model</span>
<span class="sd">      chain: If provided, returned atoms must match this chain</span>
<span class="sd">      res_type: If provided, returned atoms must match this res_type</span>

<span class="sd">    Returns:</span>
<span class="sd">      Copy of the internal dataframe that matches the input query</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">df</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">df</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
    <span class="k">if</span> <span class="n">model</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
      <span class="n">df</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">df</span><span class="o">.</span><span class="n">model</span> <span class="o">==</span> <span class="n">model</span><span class="p">]</span>
    <span class="k">if</span> <span class="n">chain</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
      <span class="n">df</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">df</span><span class="o">.</span><span class="n">chain</span> <span class="o">==</span> <span class="n">chain</span><span class="p">]</span>
    <span class="k">if</span> <span class="n">res_type</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
      <span class="n">df</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">df</span><span class="o">.</span><span class="n">res_type</span> <span class="o">==</span> <span class="n">res_type</span><span class="p">]</span>
    <span class="k">return</span> <span class="n">df</span></div>


<div class="viewcode-block" id="Protein.__sub__">
<a class="viewcode-back" href="../../neurosnap.html#neurosnap.protein.Protein.__sub__">[docs]</a>
  <span class="k">def</span><span class="w"> </span><span class="fm">__sub__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">other_protein</span><span class="p">:</span> <span class="s2">&quot;Protein&quot;</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Automatically calculate the RMSD of two proteins.</span>
<span class="sd">    Model used will naively be the first models that have</span>
<span class="sd">    identical backbone shapes.</span>
<span class="sd">    Essentially just wraps around :obj:`self.calculate_rmsd()`</span>

<span class="sd">    Parameters:</span>
<span class="sd">      other_protein: Another Protein object to compare against</span>

<span class="sd">    Returns:</span>
<span class="sd">      Copy of the internal dataframe that matches the input query</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Get models for each structure using backbone shape</span>
    <span class="n">model1</span> <span class="o">=</span> <span class="kc">None</span>
    <span class="n">model2</span> <span class="o">=</span> <span class="kc">None</span>
    <span class="k">for</span> <span class="n">m1</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">structure</span><span class="p">:</span>
      <span class="n">backbone1</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">get_backbone</span><span class="p">(</span><span class="n">model</span><span class="o">=</span><span class="n">m1</span><span class="o">.</span><span class="n">id</span><span class="p">)</span>
      <span class="k">for</span> <span class="n">m2</span> <span class="ow">in</span> <span class="n">other_protein</span><span class="o">.</span><span class="n">structure</span><span class="p">:</span>
        <span class="n">backbone2</span> <span class="o">=</span> <span class="n">other_protein</span><span class="o">.</span><span class="n">get_backbone</span><span class="p">(</span><span class="n">model</span><span class="o">=</span><span class="n">m2</span><span class="o">.</span><span class="n">id</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">backbone1</span><span class="o">.</span><span class="n">shape</span> <span class="o">==</span> <span class="n">backbone2</span><span class="o">.</span><span class="n">shape</span><span class="p">:</span>
          <span class="n">model1</span> <span class="o">=</span> <span class="n">m1</span><span class="o">.</span><span class="n">id</span>
          <span class="n">model2</span> <span class="o">=</span> <span class="n">m2</span><span class="o">.</span><span class="n">id</span>
          <span class="k">break</span>

    <span class="k">assert</span> <span class="p">(</span>
      <span class="n">model1</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span>
    <span class="p">),</span> <span class="s2">&quot;Could not find any matching matching models to calculate RMSD for. Please ensure at least two models with matching backbone shapes are provided.&quot;</span>
    <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">calculate_rmsd</span><span class="p">(</span><span class="n">other_protein</span><span class="p">,</span> <span class="n">model1</span><span class="o">=</span><span class="n">model1</span><span class="p">,</span> <span class="n">model2</span><span class="o">=</span><span class="n">model2</span><span class="p">)</span></div>


<div class="viewcode-block" id="Protein.models">
<a class="viewcode-back" href="../../neurosnap.html#neurosnap.protein.Protein.models">[docs]</a>
  <span class="k">def</span><span class="w"> </span><span class="nf">models</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">List</span><span class="p">[</span><span class="nb">int</span><span class="p">]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Returns a list of all the model names/IDs.</span>

<span class="sd">    Returns:</span>
<span class="sd">      models: Chain names/IDs found within the PDB file</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="p">[</span><span class="n">model</span><span class="o">.</span><span class="n">id</span> <span class="k">for</span> <span class="n">model</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">structure</span><span class="p">]</span></div>


<div class="viewcode-block" id="Protein.chains">
<a class="viewcode-back" href="../../neurosnap.html#neurosnap.protein.Protein.chains">[docs]</a>
  <span class="k">def</span><span class="w"> </span><span class="nf">chains</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">model</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">0</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Returns a list of all the chain names/IDs.</span>

<span class="sd">    Parameters:</span>
<span class="sd">      model: The ID of the model you want to fetch the chains of, defaults to 0</span>

<span class="sd">    Returns:</span>
<span class="sd">      Chain names/IDs found within the PDB file</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="p">[</span><span class="n">chain</span><span class="o">.</span><span class="n">id</span> <span class="k">for</span> <span class="n">chain</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">structure</span><span class="p">[</span><span class="n">model</span><span class="p">]</span> <span class="k">if</span> <span class="n">chain</span><span class="o">.</span><span class="n">id</span><span class="o">.</span><span class="n">strip</span><span class="p">()]</span></div>


<div class="viewcode-block" id="Protein.generate_df">
<a class="viewcode-back" href="../../neurosnap.html#neurosnap.protein.Protein.generate_df">[docs]</a>
  <span class="k">def</span><span class="w"> </span><span class="nf">generate_df</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Generate the biopandas-like dataframe and update the</span>
<span class="sd">    value of self.df to the new dataframe.</span>
<span class="sd">    This method should be called whenever the internal</span>
<span class="sd">    protein structure is modified or has a transformation</span>
<span class="sd">    applied to it.</span>

<span class="sd">    Inspired by: https://biopandas.github.io/biopandas</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">df</span> <span class="o">=</span> <span class="p">{</span>
      <span class="s2">&quot;model&quot;</span><span class="p">:</span> <span class="p">[],</span>
      <span class="s2">&quot;chain&quot;</span><span class="p">:</span> <span class="p">[],</span>
      <span class="s2">&quot;res_id&quot;</span><span class="p">:</span> <span class="p">[],</span>
      <span class="s2">&quot;res_name&quot;</span><span class="p">:</span> <span class="p">[],</span>
      <span class="s2">&quot;res_type&quot;</span><span class="p">:</span> <span class="p">[],</span>
      <span class="s2">&quot;atom&quot;</span><span class="p">:</span> <span class="p">[],</span>
      <span class="s2">&quot;atom_name&quot;</span><span class="p">:</span> <span class="p">[],</span>
      <span class="s2">&quot;bfactor&quot;</span><span class="p">:</span> <span class="p">[],</span>
      <span class="s2">&quot;x&quot;</span><span class="p">:</span> <span class="p">[],</span>
      <span class="s2">&quot;y&quot;</span><span class="p">:</span> <span class="p">[],</span>
      <span class="s2">&quot;z&quot;</span><span class="p">:</span> <span class="p">[],</span>
      <span class="s2">&quot;mass&quot;</span><span class="p">:</span> <span class="p">[],</span>
    <span class="p">}</span>
    <span class="k">for</span> <span class="n">model</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">structure</span><span class="p">:</span>
      <span class="k">for</span> <span class="n">chain</span> <span class="ow">in</span> <span class="n">model</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">res</span> <span class="ow">in</span> <span class="n">chain</span><span class="p">:</span>
          <span class="k">for</span> <span class="n">atom</span> <span class="ow">in</span> <span class="n">res</span><span class="p">:</span>
            <span class="c1"># get residue type</span>
            <span class="n">res_type</span> <span class="o">=</span> <span class="s2">&quot;HETEROGEN&quot;</span>
            <span class="k">if</span> <span class="n">res</span><span class="o">.</span><span class="n">id</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot; &quot;</span> <span class="ow">and</span> <span class="n">res</span><span class="o">.</span><span class="n">resname</span> <span class="ow">in</span> <span class="n">AA_ABR_TO_CODE</span><span class="p">:</span>
              <span class="n">res_type</span> <span class="o">=</span> <span class="s2">&quot;AMINO_ACID&quot;</span>
            <span class="k">elif</span> <span class="n">res</span><span class="o">.</span><span class="n">id</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot; &quot;</span> <span class="ow">and</span> <span class="n">res</span><span class="o">.</span><span class="n">resname</span> <span class="ow">in</span> <span class="n">STANDARD_NUCLEOTIDES</span><span class="p">:</span>
              <span class="n">res_type</span> <span class="o">=</span> <span class="s2">&quot;NUCLEOTIDE&quot;</span>
            <span class="n">df</span><span class="p">[</span><span class="s2">&quot;model&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">model</span><span class="o">.</span><span class="n">id</span><span class="p">)</span>
            <span class="n">df</span><span class="p">[</span><span class="s2">&quot;chain&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">chain</span><span class="o">.</span><span class="n">id</span><span class="p">)</span>
            <span class="n">df</span><span class="p">[</span><span class="s2">&quot;res_id&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">res</span><span class="o">.</span><span class="n">id</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>
            <span class="n">df</span><span class="p">[</span><span class="s2">&quot;res_name&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">res</span><span class="o">.</span><span class="n">resname</span><span class="p">)</span>
            <span class="n">df</span><span class="p">[</span><span class="s2">&quot;res_type&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">res_type</span><span class="p">)</span>
            <span class="n">df</span><span class="p">[</span><span class="s2">&quot;atom&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">atom</span><span class="o">.</span><span class="n">serial_number</span><span class="p">)</span>
            <span class="n">df</span><span class="p">[</span><span class="s2">&quot;atom_name&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">atom</span><span class="o">.</span><span class="n">name</span><span class="p">)</span>
            <span class="n">df</span><span class="p">[</span><span class="s2">&quot;bfactor&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">atom</span><span class="o">.</span><span class="n">bfactor</span><span class="p">)</span>
            <span class="n">df</span><span class="p">[</span><span class="s2">&quot;x&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">atom</span><span class="o">.</span><span class="n">coord</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
            <span class="n">df</span><span class="p">[</span><span class="s2">&quot;y&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">atom</span><span class="o">.</span><span class="n">coord</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>
            <span class="n">df</span><span class="p">[</span><span class="s2">&quot;z&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">atom</span><span class="o">.</span><span class="n">coord</span><span class="p">[</span><span class="mi">2</span><span class="p">])</span>
            <span class="n">df</span><span class="p">[</span><span class="s2">&quot;mass&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">atom</span><span class="o">.</span><span class="n">mass</span><span class="p">)</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">df</span><span class="p">)</span></div>


<div class="viewcode-block" id="Protein.get_aas">
<a class="viewcode-back" href="../../neurosnap.html#neurosnap.protein.Protein.get_aas">[docs]</a>
  <span class="k">def</span><span class="w"> </span><span class="nf">get_aas</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">model</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span> <span class="n">chain</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">str</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Returns the amino acid sequence of a target chain.</span>
<span class="sd">    Ligands, small molecules, and nucleotides are ignored.</span>

<span class="sd">    Parameters:</span>
<span class="sd">      model: The ID of the model containing the target chain</span>
<span class="sd">      chain: The ID of the chain you want to fetch the AA sequence of</span>

<span class="sd">    Returns:</span>
<span class="sd">      The amino acid sequence of the found chain</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">assert</span> <span class="n">model</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">structure</span><span class="p">,</span> <span class="sa">f</span><span class="s1">&#39;Protein does not contain model &quot;</span><span class="si">{</span><span class="n">model</span><span class="si">}</span><span class="s1">&quot;&#39;</span>
    <span class="k">assert</span> <span class="n">chain</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">structure</span><span class="p">[</span><span class="n">model</span><span class="p">],</span> <span class="sa">f</span><span class="s1">&#39;Model </span><span class="si">{</span><span class="n">model</span><span class="si">}</span><span class="s1"> does not contain chain &quot;</span><span class="si">{</span><span class="n">chain</span><span class="si">}</span><span class="s1">&quot;&#39;</span>

    <span class="n">ppb</span> <span class="o">=</span> <span class="n">PPBuilder</span><span class="p">()</span>
    <span class="k">return</span> <span class="nb">str</span><span class="p">(</span><span class="n">ppb</span><span class="o">.</span><span class="n">build_peptides</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">structure</span><span class="p">[</span><span class="n">model</span><span class="p">][</span><span class="n">chain</span><span class="p">])[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">get_sequence</span><span class="p">())</span></div>


<div class="viewcode-block" id="Protein.renumber">
<a class="viewcode-back" href="../../neurosnap.html#neurosnap.protein.Protein.renumber">[docs]</a>
  <span class="k">def</span><span class="w"> </span><span class="nf">renumber</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">model</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">int</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">chain</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">int</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">start</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">1</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Renumbers all selected residues. If selection does not</span>
<span class="sd">    exist this function will do absolutely nothing.</span>

<span class="sd">    Parameters:</span>
<span class="sd">      model: The model ID to renumber. If ``None``, will use all models.</span>
<span class="sd">      chain: The chain ID to renumber. If ``None``, will use all models.</span>
<span class="sd">      start: Starting value to increment from, defaults to 1.</span>

<span class="sd">    &quot;&quot;&quot;</span>

    <span class="k">def</span><span class="w"> </span><span class="nf">aux</span><span class="p">(</span><span class="n">start</span><span class="p">):</span>
      <span class="k">for</span> <span class="n">m</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">structure</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">model</span> <span class="ow">is</span> <span class="kc">None</span> <span class="ow">or</span> <span class="n">m</span><span class="o">.</span><span class="n">id</span> <span class="o">==</span> <span class="n">model</span><span class="p">:</span>
          <span class="k">for</span> <span class="n">c</span> <span class="ow">in</span> <span class="n">m</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">chain</span> <span class="ow">is</span> <span class="kc">None</span> <span class="ow">or</span> <span class="n">c</span><span class="o">.</span><span class="n">id</span> <span class="o">==</span> <span class="n">chain</span><span class="p">:</span>
              <span class="k">for</span> <span class="n">res</span> <span class="ow">in</span> <span class="n">c</span><span class="p">:</span>
                <span class="c1"># Renumber residue</span>
                <span class="n">res</span><span class="o">.</span><span class="n">id</span> <span class="o">=</span> <span class="p">(</span><span class="n">res</span><span class="o">.</span><span class="n">id</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="n">start</span><span class="p">,</span> <span class="n">res</span><span class="o">.</span><span class="n">id</span><span class="p">[</span><span class="mi">2</span><span class="p">])</span>
                <span class="n">start</span> <span class="o">+=</span> <span class="mi">1</span>

    <span class="c1"># perform initial renumbering to avoid collisions</span>
    <span class="n">aux</span><span class="p">(</span><span class="o">-</span><span class="mi">100000</span><span class="p">)</span>
    <span class="c1"># perform actual renumbering</span>
    <span class="n">aux</span><span class="p">(</span><span class="n">start</span><span class="p">)</span>
    <span class="c1"># update the pandas dataframe</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">generate_df</span><span class="p">()</span></div>


<div class="viewcode-block" id="Protein.remove_waters">
<a class="viewcode-back" href="../../neurosnap.html#neurosnap.protein.Protein.remove_waters">[docs]</a>
  <span class="k">def</span><span class="w"> </span><span class="nf">remove_waters</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Removes all water molecules (residues named &#39;WAT&#39; or &#39;HOH&#39;)</span>
<span class="sd">    from the structure. It is suggested to call :obj:`.renumber()`</span>
<span class="sd">    afterwards as well.</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">for</span> <span class="n">model</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">structure</span><span class="p">:</span>
      <span class="k">for</span> <span class="n">chain</span> <span class="ow">in</span> <span class="n">model</span><span class="p">:</span>
        <span class="c1"># Identify water molecules and mark them for removal</span>
        <span class="n">residues_to_remove</span> <span class="o">=</span> <span class="p">[</span><span class="n">res</span> <span class="k">for</span> <span class="n">res</span> <span class="ow">in</span> <span class="n">chain</span> <span class="k">if</span> <span class="n">res</span><span class="o">.</span><span class="n">get_resname</span><span class="p">()</span> <span class="ow">in</span> <span class="p">[</span><span class="s2">&quot;WAT&quot;</span><span class="p">,</span> <span class="s2">&quot;HOH&quot;</span><span class="p">]]</span>

        <span class="c1"># Remove water molecules</span>
        <span class="k">for</span> <span class="n">res</span> <span class="ow">in</span> <span class="n">residues_to_remove</span><span class="p">:</span>
          <span class="n">chain</span><span class="o">.</span><span class="n">detach_child</span><span class="p">(</span><span class="n">res</span><span class="o">.</span><span class="n">id</span><span class="p">)</span>
    <span class="c1"># update the pandas dataframe</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">generate_df</span><span class="p">()</span></div>


<div class="viewcode-block" id="Protein.remove_nucleotides">
<a class="viewcode-back" href="../../neurosnap.html#neurosnap.protein.Protein.remove_nucleotides">[docs]</a>
  <span class="k">def</span><span class="w"> </span><span class="nf">remove_nucleotides</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">model</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">int</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">chain</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Removes all nucleotides (DNA and RNA) from the structure.</span>
<span class="sd">    If no model or chain is provided, it will remove nucleotides</span>
<span class="sd">    from the entire structure.</span>

<span class="sd">    Parameters:</span>
<span class="sd">      model: The model ID to process. If ``None``, will use all models.</span>
<span class="sd">      chain: The chain ID to process. If ``None``, will use all chains.</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">for</span> <span class="n">m</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">structure</span><span class="p">:</span>
      <span class="k">if</span> <span class="n">model</span> <span class="ow">is</span> <span class="kc">None</span> <span class="ow">or</span> <span class="n">m</span><span class="o">.</span><span class="n">id</span> <span class="o">==</span> <span class="n">model</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">c</span> <span class="ow">in</span> <span class="n">m</span><span class="p">:</span>
          <span class="k">if</span> <span class="n">chain</span> <span class="ow">is</span> <span class="kc">None</span> <span class="ow">or</span> <span class="n">c</span><span class="o">.</span><span class="n">id</span> <span class="o">==</span> <span class="n">chain</span><span class="p">:</span>
            <span class="c1"># Identify nucleotide residues (both RNA and DNA)</span>
            <span class="n">residues_to_remove</span> <span class="o">=</span> <span class="p">[</span><span class="n">res</span> <span class="k">for</span> <span class="n">res</span> <span class="ow">in</span> <span class="n">c</span> <span class="k">if</span> <span class="n">res</span><span class="o">.</span><span class="n">get_resname</span><span class="p">()</span> <span class="ow">in</span> <span class="n">STANDARD_NUCLEOTIDES</span><span class="p">]</span>

            <span class="c1"># Remove nucleotide residues</span>
            <span class="k">for</span> <span class="n">res</span> <span class="ow">in</span> <span class="n">residues_to_remove</span><span class="p">:</span>
              <span class="n">c</span><span class="o">.</span><span class="n">detach_child</span><span class="p">(</span><span class="n">res</span><span class="o">.</span><span class="n">id</span><span class="p">)</span>

    <span class="c1"># update the pandas dataframe</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">generate_df</span><span class="p">()</span></div>


<div class="viewcode-block" id="Protein.remove_non_biopolymers">
<a class="viewcode-back" href="../../neurosnap.html#neurosnap.protein.Protein.remove_non_biopolymers">[docs]</a>
  <span class="k">def</span><span class="w"> </span><span class="nf">remove_non_biopolymers</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">model</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">int</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">chain</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Removes all ligands, heteroatoms, and non-biopolymer</span>
<span class="sd">    residues from the selected structure. Non-biopolymer</span>
<span class="sd">    residues are considered to be any residues that are not</span>
<span class="sd">    standard amino acids or standard nucleotides (DNA/RNA).</span>
<span class="sd">    If no model or chain is provided, it will remove from</span>
<span class="sd">    the entire structure.</span>

<span class="sd">    Parameters:</span>
<span class="sd">      model: The model ID to process. If ``None``, will use all models.</span>
<span class="sd">      chain: The chain ID to process. If ``None``, will use all chains.</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># List of standard amino acids and nucleotides (biopolymer residues)</span>
    <span class="n">biopolymer_residues</span> <span class="o">=</span> <span class="nb">set</span><span class="p">(</span><span class="n">AA_ABR_TO_CODE</span><span class="o">.</span><span class="n">keys</span><span class="p">())</span><span class="o">.</span><span class="n">union</span><span class="p">(</span><span class="n">STANDARD_NUCLEOTIDES</span><span class="p">)</span>
    <span class="n">biopolymer_residues</span><span class="o">.</span><span class="n">remove</span><span class="p">(</span><span class="s2">&quot;UNK&quot;</span><span class="p">)</span>

    <span class="k">for</span> <span class="n">m</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">structure</span><span class="p">:</span>
      <span class="k">if</span> <span class="n">model</span> <span class="ow">is</span> <span class="kc">None</span> <span class="ow">or</span> <span class="n">m</span><span class="o">.</span><span class="n">id</span> <span class="o">==</span> <span class="n">model</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">c</span> <span class="ow">in</span> <span class="n">m</span><span class="p">:</span>
          <span class="k">if</span> <span class="n">chain</span> <span class="ow">is</span> <span class="kc">None</span> <span class="ow">or</span> <span class="n">c</span><span class="o">.</span><span class="n">id</span> <span class="o">==</span> <span class="n">chain</span><span class="p">:</span>
            <span class="c1"># Identify non-biopolymer residues (ligands, heteroatoms, etc.)</span>
            <span class="n">residues_to_remove</span> <span class="o">=</span> <span class="p">[</span><span class="n">res</span> <span class="k">for</span> <span class="n">res</span> <span class="ow">in</span> <span class="n">c</span> <span class="k">if</span> <span class="n">res</span><span class="o">.</span><span class="n">get_resname</span><span class="p">()</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">biopolymer_residues</span><span class="p">]</span>

            <span class="c1"># Remove non-biopolymer residues</span>
            <span class="k">for</span> <span class="n">res</span> <span class="ow">in</span> <span class="n">residues_to_remove</span><span class="p">:</span>
              <span class="n">c</span><span class="o">.</span><span class="n">detach_child</span><span class="p">(</span><span class="n">res</span><span class="o">.</span><span class="n">id</span><span class="p">)</span>
    <span class="c1"># update the pandas dataframe</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">generate_df</span><span class="p">()</span></div>


<div class="viewcode-block" id="Protein.get_backbone">
<a class="viewcode-back" href="../../neurosnap.html#neurosnap.protein.Protein.get_backbone">[docs]</a>
  <span class="k">def</span><span class="w"> </span><span class="nf">get_backbone</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">model</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">int</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">chain</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Extract backbone atoms (N, CA, C) from the structure.</span>
<span class="sd">    If model or chain is not provided, extracts from all models/chains.</span>

<span class="sd">    Parameters:</span>
<span class="sd">      model: Model ID to extract from. If ``None``, all models are included.</span>
<span class="sd">      chain: Chain ID to extract from. If ``None``, all chains are included.</span>

<span class="sd">    Returns:</span>
<span class="sd">      A numpy array of backbone coordinates (Nx3)</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">backbone_coords</span> <span class="o">=</span> <span class="p">[]</span>

    <span class="k">for</span> <span class="n">m</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">structure</span><span class="p">:</span>
      <span class="k">if</span> <span class="n">model</span> <span class="ow">is</span> <span class="kc">None</span> <span class="ow">or</span> <span class="n">m</span><span class="o">.</span><span class="n">id</span> <span class="o">==</span> <span class="n">model</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">c</span> <span class="ow">in</span> <span class="n">m</span><span class="p">:</span>
          <span class="k">if</span> <span class="n">chain</span> <span class="ow">is</span> <span class="kc">None</span> <span class="ow">or</span> <span class="n">c</span><span class="o">.</span><span class="n">id</span> <span class="o">==</span> <span class="n">chain</span><span class="p">:</span>
            <span class="k">for</span> <span class="n">res</span> <span class="ow">in</span> <span class="n">c</span><span class="p">:</span>
              <span class="k">for</span> <span class="n">atom</span> <span class="ow">in</span> <span class="n">res</span><span class="p">:</span>
                <span class="k">if</span> <span class="n">atom</span><span class="o">.</span><span class="n">name</span> <span class="ow">in</span> <span class="n">BACKBONE_ATOMS</span><span class="p">:</span>
                  <span class="n">backbone_coords</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">atom</span><span class="o">.</span><span class="n">coord</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">backbone_coords</span><span class="p">)</span></div>


<div class="viewcode-block" id="Protein.find_disulfide_bonds">
<a class="viewcode-back" href="../../neurosnap.html#neurosnap.protein.Protein.find_disulfide_bonds">[docs]</a>
  <span class="k">def</span><span class="w"> </span><span class="nf">find_disulfide_bonds</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">threshold</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mf">2.05</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">List</span><span class="p">[</span><span class="n">Tuple</span><span class="p">]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Find disulfide bonds between Cysteine residues in the structure.</span>
<span class="sd">    Looks for SG-SG bonds within a threshold distance.</span>

<span class="sd">    Parameters:</span>
<span class="sd">      threshold: Maximum distance to consider a bond between SG atoms, in angstroms.</span>
<span class="sd">        Default is 2.05 .</span>

<span class="sd">    Returns:</span>
<span class="sd">      List of tuples of residue pairs forming disulfide bonds</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">disulfide_pairs</span> <span class="o">=</span> <span class="p">[]</span>

    <span class="k">for</span> <span class="n">model</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">structure</span><span class="p">:</span>
      <span class="k">for</span> <span class="n">chain</span> <span class="ow">in</span> <span class="n">model</span><span class="p">:</span>
        <span class="n">cysteines</span> <span class="o">=</span> <span class="p">[</span><span class="n">res</span> <span class="k">for</span> <span class="n">res</span> <span class="ow">in</span> <span class="n">chain</span> <span class="k">if</span> <span class="n">res</span><span class="o">.</span><span class="n">get_resname</span><span class="p">()</span> <span class="o">==</span> <span class="s2">&quot;CYS&quot;</span><span class="p">]</span>
        <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">res1</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">cysteines</span><span class="p">):</span>
          <span class="k">for</span> <span class="n">res2</span> <span class="ow">in</span> <span class="n">cysteines</span><span class="p">[</span><span class="n">i</span> <span class="o">+</span> <span class="mi">1</span> <span class="p">:]:</span>
            <span class="k">try</span><span class="p">:</span>
              <span class="n">sg1</span> <span class="o">=</span> <span class="n">res1</span><span class="p">[</span><span class="s2">&quot;SG&quot;</span><span class="p">]</span>
              <span class="n">sg2</span> <span class="o">=</span> <span class="n">res2</span><span class="p">[</span><span class="s2">&quot;SG&quot;</span><span class="p">]</span>
              <span class="n">distance</span> <span class="o">=</span> <span class="n">sg1</span> <span class="o">-</span> <span class="n">sg2</span>
              <span class="k">if</span> <span class="n">distance</span> <span class="o">&lt;</span> <span class="n">threshold</span><span class="p">:</span>
                <span class="n">disulfide_pairs</span><span class="o">.</span><span class="n">append</span><span class="p">((</span><span class="n">res1</span><span class="p">,</span> <span class="n">res2</span><span class="p">))</span>
            <span class="k">except</span> <span class="ne">KeyError</span><span class="p">:</span>
              <span class="k">pass</span>  <span class="c1"># Skip if no SG atom found</span>
    <span class="k">return</span> <span class="n">disulfide_pairs</span></div>


<div class="viewcode-block" id="Protein.find_salt_bridges">
<a class="viewcode-back" href="../../neurosnap.html#neurosnap.protein.Protein.find_salt_bridges">[docs]</a>
  <span class="k">def</span><span class="w"> </span><span class="nf">find_salt_bridges</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">model</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">int</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">chain</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">cutoff</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mf">4.0</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">List</span><span class="p">[</span><span class="n">Tuple</span><span class="p">]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Identify salt bridges between oppositely charged residues.</span>
<span class="sd">    A salt bridge is defined as an interaction between</span>
<span class="sd">    a positively charged residue (Lys, Arg) and a negatively</span>
<span class="sd">    charged residue (Asp, Glu) within a given cutoff distance.</span>

<span class="sd">    Parameters:</span>
<span class="sd">      model: Model ID to search. If ``None``, all models are searched.</span>
<span class="sd">      chain: Chain ID to search. If ``None``, all chains are searched.</span>
<span class="sd">      cutoff: Maximum distance for a salt bridge (float)</span>

<span class="sd">    Returns:</span>
<span class="sd">      List of residue pairs forming salt bridges</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">positive_residues</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;LYS&quot;</span><span class="p">,</span> <span class="s2">&quot;ARG&quot;</span><span class="p">}</span>
    <span class="n">negative_residues</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;ASP&quot;</span><span class="p">,</span> <span class="s2">&quot;GLU&quot;</span><span class="p">}</span>
    <span class="n">salt_bridges</span> <span class="o">=</span> <span class="p">[]</span>

    <span class="k">for</span> <span class="n">m</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">structure</span><span class="p">:</span>
      <span class="k">if</span> <span class="n">model</span> <span class="ow">is</span> <span class="kc">None</span> <span class="ow">or</span> <span class="n">m</span><span class="o">.</span><span class="n">id</span> <span class="o">==</span> <span class="n">model</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">c</span> <span class="ow">in</span> <span class="n">m</span><span class="p">:</span>
          <span class="k">if</span> <span class="n">chain</span> <span class="ow">is</span> <span class="kc">None</span> <span class="ow">or</span> <span class="n">c</span><span class="o">.</span><span class="n">id</span> <span class="o">==</span> <span class="n">chain</span><span class="p">:</span>
            <span class="n">pos_residues</span> <span class="o">=</span> <span class="p">[</span><span class="n">res</span> <span class="k">for</span> <span class="n">res</span> <span class="ow">in</span> <span class="n">c</span> <span class="k">if</span> <span class="n">res</span><span class="o">.</span><span class="n">get_resname</span><span class="p">()</span> <span class="ow">in</span> <span class="n">positive_residues</span><span class="p">]</span>
            <span class="n">neg_residues</span> <span class="o">=</span> <span class="p">[</span><span class="n">res</span> <span class="k">for</span> <span class="n">res</span> <span class="ow">in</span> <span class="n">c</span> <span class="k">if</span> <span class="n">res</span><span class="o">.</span><span class="n">get_resname</span><span class="p">()</span> <span class="ow">in</span> <span class="n">negative_residues</span><span class="p">]</span>
            <span class="k">for</span> <span class="n">pos_res</span> <span class="ow">in</span> <span class="n">pos_residues</span><span class="p">:</span>
              <span class="k">for</span> <span class="n">neg_res</span> <span class="ow">in</span> <span class="n">neg_residues</span><span class="p">:</span>
                <span class="n">dist</span> <span class="o">=</span> <span class="n">pos_res</span><span class="p">[</span><span class="s2">&quot;CA&quot;</span><span class="p">]</span> <span class="o">-</span> <span class="n">neg_res</span><span class="p">[</span><span class="s2">&quot;CA&quot;</span><span class="p">]</span>  <span class="c1"># Use alpha-carbon distance as a proxy</span>
                <span class="k">if</span> <span class="n">dist</span> <span class="o">&lt;</span> <span class="n">cutoff</span><span class="p">:</span>
                  <span class="n">salt_bridges</span><span class="o">.</span><span class="n">append</span><span class="p">((</span><span class="n">pos_res</span><span class="p">,</span> <span class="n">neg_res</span><span class="p">))</span>
    <span class="k">return</span> <span class="n">salt_bridges</span></div>


<div class="viewcode-block" id="Protein.find_hydrophobic_residues">
<a class="viewcode-back" href="../../neurosnap.html#neurosnap.protein.Protein.find_hydrophobic_residues">[docs]</a>
  <span class="k">def</span><span class="w"> </span><span class="nf">find_hydrophobic_residues</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">model</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">int</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">chain</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">List</span><span class="p">[</span><span class="n">Tuple</span><span class="p">]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Identify hydrophobic residues in the structure.</span>

<span class="sd">    Parameters:</span>
<span class="sd">      model: Model ID to extract from. If ``None``, all models are checked.</span>
<span class="sd">      chain: Chain ID to extract from. If ``None``, all chains are checked.</span>

<span class="sd">    Returns:</span>
<span class="sd">      List of tuples ``(model_id, chain_id, residue)`` for hydrophobic residues</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">hydrophobic_residues</span> <span class="o">=</span> <span class="p">[]</span>

    <span class="k">for</span> <span class="n">m</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">structure</span><span class="p">:</span>
      <span class="k">if</span> <span class="n">model</span> <span class="ow">is</span> <span class="kc">None</span> <span class="ow">or</span> <span class="n">m</span><span class="o">.</span><span class="n">id</span> <span class="o">==</span> <span class="n">model</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">c</span> <span class="ow">in</span> <span class="n">m</span><span class="p">:</span>
          <span class="k">if</span> <span class="n">chain</span> <span class="ow">is</span> <span class="kc">None</span> <span class="ow">or</span> <span class="n">c</span><span class="o">.</span><span class="n">id</span> <span class="o">==</span> <span class="n">chain</span><span class="p">:</span>
            <span class="k">for</span> <span class="n">res</span> <span class="ow">in</span> <span class="n">c</span><span class="p">:</span>
              <span class="k">if</span> <span class="n">res</span><span class="o">.</span><span class="n">get_resname</span><span class="p">()</span> <span class="ow">in</span> <span class="n">HYDROPHOBIC_RESIDUES</span><span class="p">:</span>
                <span class="n">hydrophobic_residues</span><span class="o">.</span><span class="n">append</span><span class="p">((</span><span class="n">m</span><span class="o">.</span><span class="n">id</span><span class="p">,</span> <span class="n">c</span><span class="o">.</span><span class="n">id</span><span class="p">,</span> <span class="n">res</span><span class="p">))</span>

    <span class="k">return</span> <span class="n">hydrophobic_residues</span></div>


<div class="viewcode-block" id="Protein.find_missing_residues">
<a class="viewcode-back" href="../../neurosnap.html#neurosnap.protein.Protein.find_missing_residues">[docs]</a>
  <span class="k">def</span><span class="w"> </span><span class="nf">find_missing_residues</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">chain</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">List</span><span class="p">[</span><span class="nb">int</span><span class="p">]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Identify missing residues in the structure based on residue numbering.</span>
<span class="sd">    Useful for identifying gaps in the structure.</span>

<span class="sd">    Parameters:</span>
<span class="sd">      chain: Chain ID to inspect. If ``None``, all chains are inspected.</span>

<span class="sd">    Returns:</span>
<span class="sd">      missing_residues: List of missing residue positions</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">missing_residues</span> <span class="o">=</span> <span class="p">[]</span>

    <span class="k">for</span> <span class="n">model</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">structure</span><span class="p">:</span>
      <span class="k">for</span> <span class="n">chain</span> <span class="ow">in</span> <span class="n">model</span><span class="p">:</span>
        <span class="n">residues</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">res</span><span class="o">.</span><span class="n">id</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="k">for</span> <span class="n">res</span> <span class="ow">in</span> <span class="n">chain</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">residues</span><span class="p">)</span> <span class="o">-</span> <span class="mi">1</span><span class="p">):</span>
          <span class="k">if</span> <span class="n">residues</span><span class="p">[</span><span class="n">i</span> <span class="o">+</span> <span class="mi">1</span><span class="p">]</span> <span class="o">!=</span> <span class="n">residues</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">+</span> <span class="mi">1</span><span class="p">:</span>
            <span class="n">missing_residues</span><span class="o">.</span><span class="n">extend</span><span class="p">(</span><span class="nb">range</span><span class="p">(</span><span class="n">residues</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">+</span> <span class="mi">1</span><span class="p">,</span> <span class="n">residues</span><span class="p">[</span><span class="n">i</span> <span class="o">+</span> <span class="mi">1</span><span class="p">]))</span>

    <span class="k">return</span> <span class="n">missing_residues</span></div>


<div class="viewcode-block" id="Protein.align">
<a class="viewcode-back" href="../../neurosnap.html#neurosnap.protein.Protein.align">[docs]</a>
  <span class="k">def</span><span class="w"> </span><span class="nf">align</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">other_protein</span><span class="p">:</span> <span class="s2">&quot;Protein&quot;</span><span class="p">,</span> <span class="n">model1</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">0</span><span class="p">,</span> <span class="n">model2</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">0</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Align another Protein object&#39;s structure to the self.structure</span>
<span class="sd">    of the current object. The other Protein will be transformed</span>
<span class="sd">    and aligned. Only compares backbone atoms (N, CA, C).</span>

<span class="sd">    Parameters:</span>
<span class="sd">      other_protein: Another Protein object to compare against</span>
<span class="sd">      model1: Model ID of reference protein to align to</span>
<span class="sd">      model2: Model ID of other protein to transform and align to reference</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">assert</span> <span class="n">model1</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">models</span><span class="p">(),</span> <span class="s2">&quot;Specified model needs to be present in the reference structure.&quot;</span>
    <span class="k">assert</span> <span class="n">model2</span> <span class="ow">in</span> <span class="n">other_protein</span><span class="o">.</span><span class="n">models</span><span class="p">(),</span> <span class="s2">&quot;Specified model needs to be present in the other structure.&quot;</span>

    <span class="c1"># Use the Superimposer to align the structures</span>
    <span class="k">def</span><span class="w"> </span><span class="nf">aux</span><span class="p">(</span><span class="n">sample_model</span><span class="p">):</span>
      <span class="n">atoms</span> <span class="o">=</span> <span class="p">[]</span>
      <span class="k">for</span> <span class="n">sample_chain</span> <span class="ow">in</span> <span class="n">sample_model</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">res</span> <span class="ow">in</span> <span class="n">sample_chain</span><span class="p">:</span>
          <span class="k">for</span> <span class="n">atom</span> <span class="ow">in</span> <span class="n">res</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">atom</span><span class="o">.</span><span class="n">name</span> <span class="ow">in</span> <span class="n">BACKBONE_ATOMS</span><span class="p">:</span>
              <span class="n">atoms</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">atom</span><span class="p">)</span>
      <span class="k">return</span> <span class="n">atoms</span>

    <span class="n">sup</span> <span class="o">=</span> <span class="n">Superimposer</span><span class="p">()</span>
    <span class="n">sup</span><span class="o">.</span><span class="n">set_atoms</span><span class="p">(</span><span class="n">aux</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">structure</span><span class="p">[</span><span class="n">model1</span><span class="p">]),</span> <span class="n">aux</span><span class="p">(</span><span class="n">other_protein</span><span class="o">.</span><span class="n">structure</span><span class="p">[</span><span class="n">model2</span><span class="p">]))</span>
    <span class="n">sup</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="n">other_protein</span><span class="o">.</span><span class="n">structure</span><span class="p">[</span><span class="n">model2</span><span class="p">])</span>  <span class="c1"># Apply the transformation to the other protein</span>
    <span class="c1"># update the pandas dataframe</span>
    <span class="n">other_protein</span><span class="o">.</span><span class="n">generate_df</span><span class="p">()</span></div>


<div class="viewcode-block" id="Protein.calculate_rmsd">
<a class="viewcode-back" href="../../neurosnap.html#neurosnap.protein.Protein.calculate_rmsd">[docs]</a>
  <span class="k">def</span><span class="w"> </span><span class="nf">calculate_rmsd</span><span class="p">(</span>
    <span class="bp">self</span><span class="p">,</span> <span class="n">other_protein</span><span class="p">:</span> <span class="s2">&quot;Protein&quot;</span><span class="p">,</span> <span class="n">model1</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">0</span><span class="p">,</span> <span class="n">model2</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">0</span><span class="p">,</span> <span class="n">chain1</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">chain2</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">align</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">True</span>
  <span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">float</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Calculate RMSD between the current structure and another protein.</span>
<span class="sd">    Only compares backbone atoms (N, CA, C). RMSD is in angstroms ().</span>

<span class="sd">    Parameters:</span>
<span class="sd">      other_protein: Another Protein object to compare against</span>
<span class="sd">      model1: Model ID of original protein to compare</span>
<span class="sd">      model2: Model ID of other protein to compare</span>
<span class="sd">      chain1: Chain ID of original protein, if not provided compares all chains</span>
<span class="sd">      chain2: Chain ID of other protein, if not provided compares all chains</span>
<span class="sd">      align: Whether to align the structures first using Superimposer</span>

<span class="sd">    Returns:</span>
<span class="sd">      The root-mean-square deviation between the two structures</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># ensure models are present</span>
    <span class="k">assert</span> <span class="n">model1</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">models</span><span class="p">(),</span> <span class="sa">f</span><span class="s2">&quot;Model </span><span class="si">{</span><span class="n">model1</span><span class="si">}</span><span class="s2"> was not found in current protein.&quot;</span>
    <span class="k">assert</span> <span class="n">model2</span> <span class="ow">in</span> <span class="n">other_protein</span><span class="o">.</span><span class="n">models</span><span class="p">(),</span> <span class="sa">f</span><span class="s2">&quot;Model </span><span class="si">{</span><span class="n">model2</span><span class="si">}</span><span class="s2"> was not found in other protein.&quot;</span>

    <span class="c1"># Get backbone coordinates of both structures</span>
    <span class="n">backbone1</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">get_backbone</span><span class="p">(</span><span class="n">model</span><span class="o">=</span><span class="n">model1</span><span class="p">,</span> <span class="n">chain</span><span class="o">=</span><span class="n">chain1</span><span class="p">)</span>
    <span class="n">backbone2</span> <span class="o">=</span> <span class="n">other_protein</span><span class="o">.</span><span class="n">get_backbone</span><span class="p">(</span><span class="n">model</span><span class="o">=</span><span class="n">model2</span><span class="p">,</span> <span class="n">chain</span><span class="o">=</span><span class="n">chain2</span><span class="p">)</span>
    <span class="k">assert</span> <span class="n">backbone1</span><span class="o">.</span><span class="n">shape</span> <span class="o">==</span> <span class="n">backbone2</span><span class="o">.</span><span class="n">shape</span><span class="p">,</span> <span class="s2">&quot;Structures must have the same number of backbone atoms for RMSD calculation.&quot;</span>

    <span class="k">if</span> <span class="n">align</span><span class="p">:</span>
      <span class="bp">self</span><span class="o">.</span><span class="n">align</span><span class="p">(</span><span class="n">other_protein</span><span class="p">,</span> <span class="n">model1</span><span class="o">=</span><span class="n">model1</span><span class="p">,</span> <span class="n">model2</span><span class="o">=</span><span class="n">model2</span><span class="p">)</span>

    <span class="c1"># Get new backbone coordinates of both structures</span>
    <span class="n">backbone1</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">get_backbone</span><span class="p">(</span><span class="n">model</span><span class="o">=</span><span class="n">model1</span><span class="p">,</span> <span class="n">chain</span><span class="o">=</span><span class="n">chain1</span><span class="p">)</span>
    <span class="n">backbone2</span> <span class="o">=</span> <span class="n">other_protein</span><span class="o">.</span><span class="n">get_backbone</span><span class="p">(</span><span class="n">model</span><span class="o">=</span><span class="n">model2</span><span class="p">,</span> <span class="n">chain</span><span class="o">=</span><span class="n">chain2</span><span class="p">)</span>

    <span class="n">diff</span> <span class="o">=</span> <span class="n">backbone1</span> <span class="o">-</span> <span class="n">backbone2</span>
    <span class="n">rmsd</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">sqrt</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">diff</span><span class="o">**</span><span class="mi">2</span><span class="p">)</span> <span class="o">/</span> <span class="n">backbone1</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
    <span class="k">return</span> <span class="n">rmsd</span></div>


<div class="viewcode-block" id="Protein.calculate_distance_matrix">
<a class="viewcode-back" href="../../neurosnap.html#neurosnap.protein.Protein.calculate_distance_matrix">[docs]</a>
  <span class="k">def</span><span class="w"> </span><span class="nf">calculate_distance_matrix</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">model</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">int</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">chain</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Calculate the distance matrix for all alpha-carbon (CA) atoms in the chain.</span>
<span class="sd">    Useful for creating contact maps or proximity analyses.</span>

<span class="sd">    Parameters:</span>
<span class="sd">      model: The model ID to calculate the distance matrix for, if not provided will use first model found</span>
<span class="sd">      chain: The chain ID to calculate, if not provided calculates for all chains</span>

<span class="sd">    Returns:</span>
<span class="sd">      A 2D numpy array representing the distance matrix</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">model</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">models</span><span class="p">()[</span><span class="mi">0</span><span class="p">]</span>
    <span class="n">ca_atoms</span> <span class="o">=</span> <span class="p">[]</span>

    <span class="k">for</span> <span class="n">m</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">structure</span><span class="p">:</span>
      <span class="k">if</span> <span class="n">m</span><span class="o">.</span><span class="n">id</span> <span class="o">==</span> <span class="n">model</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">c</span> <span class="ow">in</span> <span class="n">m</span><span class="p">:</span>
          <span class="k">if</span> <span class="n">chain</span> <span class="ow">is</span> <span class="kc">None</span> <span class="ow">or</span> <span class="n">c</span><span class="o">.</span><span class="n">id</span> <span class="o">==</span> <span class="n">chain</span><span class="p">:</span>
            <span class="k">for</span> <span class="n">res</span> <span class="ow">in</span> <span class="n">c</span><span class="p">:</span>
              <span class="k">if</span> <span class="s2">&quot;CA&quot;</span> <span class="ow">in</span> <span class="n">res</span><span class="p">:</span>
                <span class="n">ca_atoms</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">res</span><span class="p">[</span><span class="s2">&quot;CA&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">coord</span><span class="p">)</span>

    <span class="n">ca_atoms</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">ca_atoms</span><span class="p">)</span>
    <span class="n">dist_matrix</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">sqrt</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">sum</span><span class="p">((</span><span class="n">ca_atoms</span><span class="p">[:,</span> <span class="n">np</span><span class="o">.</span><span class="n">newaxis</span><span class="p">]</span> <span class="o">-</span> <span class="n">ca_atoms</span><span class="p">[</span><span class="n">np</span><span class="o">.</span><span class="n">newaxis</span><span class="p">,</span> <span class="p">:])</span> <span class="o">**</span> <span class="mi">2</span><span class="p">,</span> <span class="n">axis</span><span class="o">=-</span><span class="mi">1</span><span class="p">))</span>
    <span class="k">return</span> <span class="n">dist_matrix</span></div>


<div class="viewcode-block" id="Protein.calculate_center_of_mass">
<a class="viewcode-back" href="../../neurosnap.html#neurosnap.protein.Protein.calculate_center_of_mass">[docs]</a>
  <span class="k">def</span><span class="w"> </span><span class="nf">calculate_center_of_mass</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">model</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">int</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">chain</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Calculate the center of mass of the protein.</span>
<span class="sd">    Considers only atoms with defined masses.</span>

<span class="sd">    Parameters:</span>
<span class="sd">      model: Model ID to calculate for, if not provided calculates for all models</span>
<span class="sd">      chain: Chain ID to calculate for, if not provided calculates for all chains</span>

<span class="sd">    Returns:</span>
<span class="sd">      center_of_mass: A 3D numpy array representing the center of mass</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">total_mass</span> <span class="o">=</span> <span class="mi">0</span>
    <span class="n">weighted_coords</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">(</span><span class="mi">3</span><span class="p">)</span>

    <span class="k">for</span> <span class="n">m</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">structure</span><span class="p">:</span>
      <span class="k">if</span> <span class="n">model</span> <span class="ow">is</span> <span class="kc">None</span> <span class="ow">or</span> <span class="n">m</span><span class="o">.</span><span class="n">id</span> <span class="o">==</span> <span class="n">model</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">c</span> <span class="ow">in</span> <span class="n">m</span><span class="p">:</span>
          <span class="k">if</span> <span class="n">chain</span> <span class="ow">is</span> <span class="kc">None</span> <span class="ow">or</span> <span class="n">c</span><span class="o">.</span><span class="n">id</span> <span class="o">==</span> <span class="n">chain</span><span class="p">:</span>
            <span class="k">for</span> <span class="n">res</span> <span class="ow">in</span> <span class="n">c</span><span class="p">:</span>
              <span class="k">for</span> <span class="n">atom</span> <span class="ow">in</span> <span class="n">res</span><span class="p">:</span>
                <span class="k">if</span> <span class="n">atom</span><span class="o">.</span><span class="n">mass</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
                  <span class="n">total_mass</span> <span class="o">+=</span> <span class="n">atom</span><span class="o">.</span><span class="n">mass</span>
                  <span class="n">weighted_coords</span> <span class="o">+=</span> <span class="n">atom</span><span class="o">.</span><span class="n">mass</span> <span class="o">*</span> <span class="n">atom</span><span class="o">.</span><span class="n">coord</span>

    <span class="k">if</span> <span class="n">total_mass</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
      <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;No atoms with mass found in the selected structure.&quot;</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">weighted_coords</span> <span class="o">/</span> <span class="n">total_mass</span></div>


<div class="viewcode-block" id="Protein.distances_from_com">
<a class="viewcode-back" href="../../neurosnap.html#neurosnap.protein.Protein.distances_from_com">[docs]</a>
  <span class="k">def</span><span class="w"> </span><span class="nf">distances_from_com</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">model</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">int</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">chain</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Calculate the distances of all atoms from the center of mass (COM) of the protein.</span>

<span class="sd">    This method computes the Euclidean distance between the coordinates of each atom</span>
<span class="sd">    and the center of mass of the structure. The center of mass is calculated for the</span>
<span class="sd">    specified model and chain, or for all models and chains if none are provided.</span>

<span class="sd">    Parameters:</span>
<span class="sd">      model: The model ID to calculate for. If not provided, calculates for all models.</span>
<span class="sd">      chain: The chain ID to calculate for. If not provided, calculates for all chains.</span>

<span class="sd">    Returns:</span>
<span class="sd">      A 1D NumPy array containing the distances (in ngstrms) between each atom and the center of mass.</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">com</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">calculate_center_of_mass</span><span class="p">(</span><span class="n">model</span><span class="o">=</span><span class="n">model</span><span class="p">,</span> <span class="n">chain</span><span class="o">=</span><span class="n">chain</span><span class="p">)</span>
    <span class="n">distances</span> <span class="o">=</span> <span class="p">[]</span>

    <span class="k">for</span> <span class="n">m</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">structure</span><span class="p">:</span>
      <span class="k">if</span> <span class="n">model</span> <span class="ow">is</span> <span class="kc">None</span> <span class="ow">or</span> <span class="n">m</span><span class="o">.</span><span class="n">id</span> <span class="o">==</span> <span class="n">model</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">c</span> <span class="ow">in</span> <span class="n">m</span><span class="p">:</span>
          <span class="k">if</span> <span class="n">chain</span> <span class="ow">is</span> <span class="kc">None</span> <span class="ow">or</span> <span class="n">c</span><span class="o">.</span><span class="n">id</span> <span class="o">==</span> <span class="n">chain</span><span class="p">:</span>
            <span class="k">for</span> <span class="n">res</span> <span class="ow">in</span> <span class="n">c</span><span class="p">:</span>
              <span class="k">for</span> <span class="n">atom</span> <span class="ow">in</span> <span class="n">res</span><span class="p">:</span>
                <span class="n">distance</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">linalg</span><span class="o">.</span><span class="n">norm</span><span class="p">(</span><span class="n">atom</span><span class="o">.</span><span class="n">coord</span> <span class="o">-</span> <span class="n">com</span><span class="p">)</span>
                <span class="n">distances</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">distance</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">distances</span><span class="p">)</span>  <span class="c1"># Convert the list of distances to a NumPy array</span></div>


<div class="viewcode-block" id="Protein.calculate_surface_area">
<a class="viewcode-back" href="../../neurosnap.html#neurosnap.protein.Protein.calculate_surface_area">[docs]</a>
  <span class="k">def</span><span class="w"> </span><span class="nf">calculate_surface_area</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">model</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">0</span><span class="p">,</span> <span class="n">level</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;R&quot;</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">float</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Calculate the solvent-accessible surface area (SASA) of the protein.</span>
<span class="sd">    Utilizes Biopython&#39;s SASA module.</span>

<span class="sd">    Parameters:</span>
<span class="sd">      model: The model ID to calculate SASA for, defaults to 0.</span>
<span class="sd">      level: The level at which ASA values are assigned, which can be one of &quot;A&quot; (Atom), &quot;R&quot; (Residue), &quot;C&quot; (Chain), &quot;M&quot; (Model), or &quot;S&quot; (Structure). The ASA value of an entity is the sum of all ASA values of its children.</span>

<span class="sd">    Returns:</span>
<span class="sd">      Solvent-accessible surface area in </span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">assert</span> <span class="n">model</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">models</span><span class="p">(),</span> <span class="sa">f</span><span class="s2">&quot;Model </span><span class="si">{</span><span class="n">model</span><span class="si">}</span><span class="s2"> is not currently present.&quot;</span>
    <span class="n">structure_model</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">structure</span><span class="p">[</span><span class="n">model</span><span class="p">]</span>
    <span class="n">sasa_calculator</span> <span class="o">=</span> <span class="n">SASA</span><span class="o">.</span><span class="n">ShrakeRupley</span><span class="p">()</span>
    <span class="n">sasa_calculator</span><span class="o">.</span><span class="n">compute</span><span class="p">(</span><span class="n">structure_model</span><span class="p">,</span> <span class="n">level</span><span class="o">=</span><span class="n">level</span><span class="p">)</span>
    <span class="n">total_sasa</span> <span class="o">=</span> <span class="nb">sum</span><span class="p">([</span><span class="n">residue</span><span class="o">.</span><span class="n">sasa</span> <span class="k">for</span> <span class="n">residue</span> <span class="ow">in</span> <span class="n">structure_model</span><span class="o">.</span><span class="n">get_residues</span><span class="p">()</span> <span class="k">if</span> <span class="n">residue</span><span class="o">.</span><span class="n">sasa</span><span class="p">])</span>
    <span class="k">return</span> <span class="n">total_sasa</span></div>


<div class="viewcode-block" id="Protein.calculate_protein_volume">
<a class="viewcode-back" href="../../neurosnap.html#neurosnap.protein.Protein.calculate_protein_volume">[docs]</a>
  <span class="k">def</span><span class="w"> </span><span class="nf">calculate_protein_volume</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">model</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">0</span><span class="p">,</span> <span class="n">chain</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">float</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Compute an estimate of the protein volume using the van der Waals radii.</span>
<span class="sd">    Uses the sum of atom radii to compute the volume.</span>

<span class="sd">    Parameters:</span>
<span class="sd">      model: Model ID to compute volume for, defaults to 0</span>
<span class="sd">      chain: Chain ID to compute, if not provided computes for all chains</span>

<span class="sd">    Returns:</span>
<span class="sd">      Estimated volume in </span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">assert</span> <span class="n">model</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">models</span><span class="p">(),</span> <span class="sa">f</span><span class="s2">&quot;Model </span><span class="si">{</span><span class="n">model</span><span class="si">}</span><span class="s2"> is not currently present.&quot;</span>
    <span class="n">vdw_radii</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;H&quot;</span><span class="p">:</span> <span class="mf">1.2</span><span class="p">,</span> <span class="s2">&quot;C&quot;</span><span class="p">:</span> <span class="mf">1.7</span><span class="p">,</span> <span class="s2">&quot;N&quot;</span><span class="p">:</span> <span class="mf">1.55</span><span class="p">,</span> <span class="s2">&quot;O&quot;</span><span class="p">:</span> <span class="mf">1.52</span><span class="p">,</span> <span class="s2">&quot;P&quot;</span><span class="p">:</span> <span class="mf">1.8</span><span class="p">,</span> <span class="s2">&quot;S&quot;</span><span class="p">:</span> <span class="mf">1.8</span><span class="p">}</span>  <span class="c1"># Example radii in </span>
    <span class="n">volume</span> <span class="o">=</span> <span class="mi">0</span>

    <span class="k">for</span> <span class="n">m</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">structure</span><span class="p">:</span>
      <span class="k">for</span> <span class="n">c</span> <span class="ow">in</span> <span class="n">m</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">chain</span> <span class="ow">is</span> <span class="kc">None</span> <span class="ow">or</span> <span class="n">c</span><span class="o">.</span><span class="n">id</span> <span class="o">==</span> <span class="n">chain</span><span class="p">:</span>
          <span class="k">for</span> <span class="n">res</span> <span class="ow">in</span> <span class="n">c</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">is_aa</span><span class="p">(</span><span class="n">res</span><span class="p">):</span>
              <span class="k">for</span> <span class="n">atom</span> <span class="ow">in</span> <span class="n">res</span><span class="p">:</span>
                <span class="n">element</span> <span class="o">=</span> <span class="n">atom</span><span class="o">.</span><span class="n">element</span>
                <span class="k">if</span> <span class="n">element</span> <span class="ow">in</span> <span class="n">vdw_radii</span><span class="p">:</span>
                  <span class="n">radius</span> <span class="o">=</span> <span class="n">vdw_radii</span><span class="p">[</span><span class="n">element</span><span class="p">]</span>
                  <span class="n">volume</span> <span class="o">+=</span> <span class="p">(</span><span class="mi">4</span> <span class="o">/</span> <span class="mi">3</span><span class="p">)</span> <span class="o">*</span> <span class="n">np</span><span class="o">.</span><span class="n">pi</span> <span class="o">*</span> <span class="p">(</span><span class="n">radius</span><span class="o">**</span><span class="mi">3</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">volume</span></div>


<div class="viewcode-block" id="Protein.calculate_hydrogen_bonds">
<a class="viewcode-back" href="../../neurosnap.html#neurosnap.protein.Protein.calculate_hydrogen_bonds">[docs]</a>
  <span class="k">def</span><span class="w"> </span><span class="nf">calculate_hydrogen_bonds</span><span class="p">(</span>
    <span class="bp">self</span><span class="p">,</span>
    <span class="n">model</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">int</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">chain</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">chain_other</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">donor_acceptor_cutoff</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mf">3.5</span><span class="p">,</span>
    <span class="n">angle_cutoff</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mf">120.0</span><span class="p">,</span>
  <span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">int</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Calculate the number of hydrogen bonds in the protein structure.</span>
<span class="sd">    Hydrogen atoms must be explicitly defined within the structure as implicit hydrogens</span>
<span class="sd">    will not computed. We recommend using a tool like reduce to add missing hydrogens.</span>

<span class="sd">    Hydrogen bonds are detected based on distance and angle criteria:</span>
<span class="sd">    - Distance between donor and acceptor must be less than `donor_acceptor_cutoff`.</span>
<span class="sd">    - The angle formed by donor-hydrogen-acceptor must be greater than `angle_cutoff`.</span>

<span class="sd">    If `model` is set to `None`, hydrogen bonds are calculated only for the first model in the structure.</span>

<span class="sd">    If `chain_other` is `None`:</span>
<span class="sd">      - Hydrogen bonds are calculated for the specified `chain` or all chains if `chain` is also `None`.</span>
<span class="sd">    If `chain_other` is set to a specific chain:</span>
<span class="sd">      - Hydrogen bonds are calculated only between atoms of `chain` and `chain_other`.</span>
<span class="sd">    If `chain_other` is specified but `chain` is not, an exception is raised.</span>

<span class="sd">    Parameters:</span>
<span class="sd">      model: Model ID to calculate for. If None, only the first model is considered.</span>
<span class="sd">      chain: Chain ID to calculate for. If None, all chains in the selected model are considered.</span>
<span class="sd">      chain_other: Secondary chain ID for inter-chain hydrogen bonds. If None, intra-chain bonds are calculated.</span>
<span class="sd">      donor_acceptor_cutoff: Maximum distance between donor and acceptor (in ). Default is 3.5 .</span>
<span class="sd">      angle_cutoff: Minimum angle for a hydrogen bond (in degrees). Default is 120.</span>

<span class="sd">    Returns:</span>
<span class="sd">      The total number of hydrogen bonds in the structure.</span>

<span class="sd">    Raises:</span>
<span class="sd">      ValueError: If `chain_other` is specified but `chain` is not.</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">hydrogen_bonds</span> <span class="o">=</span> <span class="mi">0</span>
    <span class="n">hydrogen_distance_cutoff</span> <span class="o">=</span> <span class="mf">1.2</span>  <span class="c1"># Typical bond length between H and donor atom (in )</span>

    <span class="c1"># Default to the first model if model is None</span>
    <span class="k">if</span> <span class="n">model</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
      <span class="n">model</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">models</span><span class="p">()[</span><span class="mi">0</span><span class="p">]</span>
    <span class="n">model</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">structure</span><span class="p">[</span><span class="n">model</span><span class="p">]</span>

    <span class="c1"># input validation</span>
    <span class="k">if</span> <span class="n">chain_other</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="ow">and</span> <span class="n">chain</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
      <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;`chain_other` is specified, but `chain` is not. Both must be provided for inter-chain hydrogen bond calculation.&quot;</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">chain</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="ow">and</span> <span class="n">chain</span> <span class="ow">not</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">chains</span><span class="p">():</span>
      <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Chain </span><span class="si">{</span><span class="n">chain</span><span class="si">}</span><span class="s2"> does not exist within the input structure.&quot;</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">chain_other</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="ow">and</span> <span class="n">chain_other</span> <span class="ow">not</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">chains</span><span class="p">():</span>
      <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Chain </span><span class="si">{</span><span class="n">chain_other</span><span class="si">}</span><span class="s2"> does not exist within the input structure.&quot;</span><span class="p">)</span>

    <span class="k">for</span> <span class="n">c</span> <span class="ow">in</span> <span class="n">model</span><span class="p">:</span>
      <span class="k">if</span> <span class="n">chain</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="ow">and</span> <span class="n">c</span><span class="o">.</span><span class="n">id</span> <span class="o">!=</span> <span class="n">chain</span><span class="p">:</span>
        <span class="k">continue</span>

      <span class="k">for</span> <span class="n">donor_res</span> <span class="ow">in</span> <span class="n">c</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">donor_atom</span> <span class="ow">in</span> <span class="n">donor_res</span><span class="p">:</span>
          <span class="k">if</span> <span class="n">donor_atom</span><span class="o">.</span><span class="n">element</span> <span class="ow">not</span> <span class="ow">in</span> <span class="p">[</span><span class="s2">&quot;N&quot;</span><span class="p">,</span> <span class="s2">&quot;O&quot;</span><span class="p">]:</span>
            <span class="k">continue</span>  <span class="c1"># Only consider N or O as donors</span>

          <span class="c1"># Identify hydrogen atoms bonded to the donor</span>
          <span class="n">bonded_hydrogens</span> <span class="o">=</span> <span class="p">[</span>
            <span class="n">atom</span> <span class="k">for</span> <span class="n">atom</span> <span class="ow">in</span> <span class="n">donor_res</span> <span class="k">if</span> <span class="n">atom</span><span class="o">.</span><span class="n">element</span> <span class="o">==</span> <span class="s2">&quot;H&quot;</span> <span class="ow">and</span> <span class="n">np</span><span class="o">.</span><span class="n">linalg</span><span class="o">.</span><span class="n">norm</span><span class="p">(</span><span class="n">donor_atom</span><span class="o">.</span><span class="n">coord</span> <span class="o">-</span> <span class="n">atom</span><span class="o">.</span><span class="n">coord</span><span class="p">)</span> <span class="o">&lt;=</span> <span class="n">hydrogen_distance_cutoff</span>
          <span class="p">]</span>
          <span class="k">if</span> <span class="ow">not</span> <span class="n">bonded_hydrogens</span><span class="p">:</span>
            <span class="k">continue</span>

          <span class="c1"># Determine chains to search for acceptors</span>
          <span class="n">acceptor_chains</span> <span class="o">=</span> <span class="p">[</span><span class="n">chain_other</span><span class="p">]</span> <span class="k">if</span> <span class="n">chain_other</span> <span class="k">else</span> <span class="p">[</span><span class="n">c</span><span class="o">.</span><span class="n">id</span> <span class="k">for</span> <span class="n">c</span> <span class="ow">in</span> <span class="n">model</span> <span class="k">if</span> <span class="n">chain</span> <span class="ow">is</span> <span class="kc">None</span> <span class="ow">or</span> <span class="n">c</span><span class="o">.</span><span class="n">id</span> <span class="o">==</span> <span class="n">chain</span><span class="p">]</span>

          <span class="k">for</span> <span class="n">acceptor_chain</span> <span class="ow">in</span> <span class="n">acceptor_chains</span><span class="p">:</span>
            <span class="k">for</span> <span class="n">acceptor_res</span> <span class="ow">in</span> <span class="n">model</span><span class="p">[</span><span class="n">acceptor_chain</span><span class="p">]:</span>
              <span class="k">for</span> <span class="n">acceptor_atom</span> <span class="ow">in</span> <span class="n">acceptor_res</span><span class="p">:</span>
                <span class="k">if</span> <span class="n">acceptor_atom</span><span class="o">.</span><span class="n">element</span> <span class="ow">not</span> <span class="ow">in</span> <span class="p">[</span><span class="s2">&quot;N&quot;</span><span class="p">,</span> <span class="s2">&quot;O&quot;</span><span class="p">]:</span>
                  <span class="k">continue</span>  <span class="c1"># Only consider N or O as acceptors</span>
                <span class="k">if</span> <span class="n">acceptor_atom</span> <span class="ow">is</span> <span class="n">donor_atom</span><span class="p">:</span>
                  <span class="k">continue</span>  <span class="c1"># Skip self-bonding</span>

                <span class="c1"># Calculate the distance between donor and acceptor</span>
                <span class="n">distance</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">linalg</span><span class="o">.</span><span class="n">norm</span><span class="p">(</span><span class="n">donor_atom</span><span class="o">.</span><span class="n">coord</span> <span class="o">-</span> <span class="n">acceptor_atom</span><span class="o">.</span><span class="n">coord</span><span class="p">)</span>
                <span class="k">if</span> <span class="n">distance</span> <span class="o">&gt;</span> <span class="n">donor_acceptor_cutoff</span><span class="p">:</span>
                  <span class="k">continue</span>

                <span class="c1"># Calculate the angle formed by donor-hydrogen-acceptor for each bonded hydrogen</span>
                <span class="k">for</span> <span class="n">hydrogen</span> <span class="ow">in</span> <span class="n">bonded_hydrogens</span><span class="p">:</span>
                  <span class="n">donor_h_vector</span> <span class="o">=</span> <span class="n">hydrogen</span><span class="o">.</span><span class="n">coord</span> <span class="o">-</span> <span class="n">donor_atom</span><span class="o">.</span><span class="n">coord</span>
                  <span class="n">acceptor_donor_vector</span> <span class="o">=</span> <span class="n">acceptor_atom</span><span class="o">.</span><span class="n">coord</span> <span class="o">-</span> <span class="n">donor_atom</span><span class="o">.</span><span class="n">coord</span>
                  <span class="n">angle</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">arccos</span><span class="p">(</span>
                    <span class="n">np</span><span class="o">.</span><span class="n">dot</span><span class="p">(</span><span class="n">donor_h_vector</span><span class="p">,</span> <span class="n">acceptor_donor_vector</span><span class="p">)</span> <span class="o">/</span> <span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">linalg</span><span class="o">.</span><span class="n">norm</span><span class="p">(</span><span class="n">donor_h_vector</span><span class="p">)</span> <span class="o">*</span> <span class="n">np</span><span class="o">.</span><span class="n">linalg</span><span class="o">.</span><span class="n">norm</span><span class="p">(</span><span class="n">acceptor_donor_vector</span><span class="p">))</span>
                  <span class="p">)</span>
                  <span class="n">angle</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">degrees</span><span class="p">(</span><span class="n">angle</span><span class="p">)</span>

                  <span class="k">if</span> <span class="n">angle</span> <span class="o">&gt;=</span> <span class="n">angle_cutoff</span><span class="p">:</span>
                    <span class="n">hydrogen_bonds</span> <span class="o">+=</span> <span class="mi">1</span>

    <span class="k">return</span> <span class="n">hydrogen_bonds</span></div>


<div class="viewcode-block" id="Protein.to_sdf">
<a class="viewcode-back" href="../../neurosnap.html#neurosnap.protein.Protein.to_sdf">[docs]</a>
  <span class="k">def</span><span class="w"> </span><span class="nf">to_sdf</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">fpath</span><span class="p">:</span> <span class="nb">str</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Save the current protein structure as an SDF file.</span>
<span class="sd">    Will export all models and chains. Use :obj:`.remove()`</span>
<span class="sd">    method to get rid of undesired regions.</span>

<span class="sd">    Parameters:</span>
<span class="sd">      fpath: Path to the output SDF file</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Write the current protein structure to a temporary PDB file</span>
    <span class="k">with</span> <span class="n">tempfile</span><span class="o">.</span><span class="n">NamedTemporaryFile</span><span class="p">(</span><span class="n">delete</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">suffix</span><span class="o">=</span><span class="s2">&quot;.pdb&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">temp_pdb</span><span class="p">:</span>
      <span class="bp">self</span><span class="o">.</span><span class="n">save</span><span class="p">(</span><span class="n">temp_pdb</span><span class="o">.</span><span class="n">name</span><span class="p">,</span> <span class="nb">format</span><span class="o">=</span><span class="s2">&quot;pdb&quot;</span><span class="p">)</span>
      <span class="n">pdb_fpath</span> <span class="o">=</span> <span class="n">temp_pdb</span><span class="o">.</span><span class="n">name</span>

      <span class="c1"># Read the PDB file</span>
      <span class="n">mol</span> <span class="o">=</span> <span class="n">Chem</span><span class="o">.</span><span class="n">MolFromPDBFile</span><span class="p">(</span><span class="n">pdb_fpath</span><span class="p">,</span> <span class="n">sanitize</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>

      <span class="k">if</span> <span class="n">mol</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;Unable to parse the PDB file.&quot;</span><span class="p">)</span>

      <span class="c1"># Write the molecule to SDF</span>
      <span class="n">writer</span> <span class="o">=</span> <span class="n">Chem</span><span class="o">.</span><span class="n">SDWriter</span><span class="p">(</span><span class="n">fpath</span><span class="p">)</span>
      <span class="n">writer</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">mol</span><span class="p">)</span>
      <span class="n">writer</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
      <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Successfully wrote SDF file to </span><span class="si">{</span><span class="n">fpath</span><span class="si">}</span><span class="s2">.&quot;</span><span class="p">)</span></div>


<div class="viewcode-block" id="Protein.remove">
<a class="viewcode-back" href="../../neurosnap.html#neurosnap.protein.Protein.remove">[docs]</a>
  <span class="k">def</span><span class="w"> </span><span class="nf">remove</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">model</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span> <span class="n">chain</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">resi_start</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">int</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">resi_end</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">int</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Completely removes all parts of a selection from</span>
<span class="sd">    self.structure. If a residue range is provided then all</span>
<span class="sd">    residues between resi_start and resi_end will be removed</span>
<span class="sd">    from the structure (inclusively). If a residue range is</span>
<span class="sd">    not provided then all residues in a chain will be removed.</span>

<span class="sd">    Parameters:</span>
<span class="sd">      model: ID of model to remove from</span>
<span class="sd">      chain: ID of chain to remove from, if not provided will remove all chains in the model</span>
<span class="sd">      resi_start: Index of first residue in the range you want to remove</span>
<span class="sd">      resi_end: Index of last residues in the range you want to remove</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># validate input query</span>
    <span class="k">assert</span> <span class="n">model</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">models</span><span class="p">(),</span> <span class="sa">f</span><span class="s2">&quot;Model ID </span><span class="si">{</span><span class="n">model</span><span class="si">}</span><span class="s2"> does not exist in your structure. Found models include </span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">models</span><span class="p">()</span><span class="si">}</span><span class="s2">.&quot;</span>
    <span class="k">if</span> <span class="n">chain</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
      <span class="k">assert</span> <span class="n">chain</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">chains</span><span class="p">(</span><span class="n">model</span><span class="p">),</span> <span class="sa">f</span><span class="s2">&quot;Chain ID </span><span class="si">{</span><span class="n">chain</span><span class="si">}</span><span class="s2"> does not exist in your structure. Found chains include </span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">chains</span><span class="p">(</span><span class="n">model</span><span class="p">)</span><span class="si">}</span><span class="s2">.&quot;</span>
    <span class="k">if</span> <span class="n">resi_start</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="ow">or</span> <span class="n">resi_end</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
      <span class="k">assert</span> <span class="n">chain</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">,</span> <span class="s2">&quot;Chain needs to specified if you want to remove residues&quot;</span>
      <span class="k">assert</span> <span class="n">resi_start</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="ow">and</span> <span class="n">resi_end</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">,</span> <span class="s2">&quot;Both resi_start and resi_end must be provided&quot;</span>
      <span class="k">assert</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">resi_start</span><span class="p">,</span> <span class="nb">int</span><span class="p">)</span> <span class="ow">and</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">resi_end</span><span class="p">,</span> <span class="nb">int</span><span class="p">),</span> <span class="s2">&quot;Both resi_start and resi_end must be valid integers&quot;</span>
      <span class="k">assert</span> <span class="n">resi_end</span> <span class="o">&gt;=</span> <span class="n">resi_start</span><span class="p">,</span> <span class="s2">&quot;resi_start start must be less than resi_end&quot;</span>
      <span class="k">assert</span> <span class="n">resi_start</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">structure</span><span class="p">[</span><span class="n">model</span><span class="p">][</span><span class="n">chain</span><span class="p">],</span> <span class="sa">f</span><span class="s2">&quot;Residue </span><span class="si">{</span><span class="n">resi_start</span><span class="si">}</span><span class="s2"> does not exist in the specified part of your structure.&quot;</span>
      <span class="k">assert</span> <span class="n">resi_end</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">structure</span><span class="p">[</span><span class="n">model</span><span class="p">][</span><span class="n">chain</span><span class="p">],</span> <span class="sa">f</span><span class="s2">&quot;Residue </span><span class="si">{</span><span class="n">resi_end</span><span class="si">}</span><span class="s2"> does not exist in the specified part of your structure.&quot;</span>

    <span class="c1"># Perform the removal</span>
    <span class="k">if</span> <span class="n">resi_start</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="ow">and</span> <span class="n">resi_end</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
      <span class="c1"># Remove residues in the specified range</span>
      <span class="n">chain_obj</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">structure</span><span class="p">[</span><span class="n">model</span><span class="p">][</span><span class="n">chain</span><span class="p">]</span>
      <span class="n">residues_to_remove</span> <span class="o">=</span> <span class="p">[</span><span class="n">res</span> <span class="k">for</span> <span class="n">res</span> <span class="ow">in</span> <span class="n">chain_obj</span> <span class="k">if</span> <span class="n">resi_start</span> <span class="o">&lt;=</span> <span class="n">res</span><span class="o">.</span><span class="n">id</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="o">&lt;=</span> <span class="n">resi_end</span><span class="p">]</span>
      <span class="k">for</span> <span class="n">res</span> <span class="ow">in</span> <span class="n">residues_to_remove</span><span class="p">:</span>
        <span class="n">chain_obj</span><span class="o">.</span><span class="n">detach_child</span><span class="p">(</span><span class="n">res</span><span class="o">.</span><span class="n">id</span><span class="p">)</span>
    <span class="k">elif</span> <span class="n">chain</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
      <span class="c1"># Remove the entire chain</span>
      <span class="bp">self</span><span class="o">.</span><span class="n">structure</span><span class="p">[</span><span class="n">model</span><span class="p">]</span><span class="o">.</span><span class="n">detach_child</span><span class="p">(</span><span class="n">chain</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
      <span class="c1"># Remove the entire model</span>
      <span class="bp">self</span><span class="o">.</span><span class="n">structure</span><span class="o">.</span><span class="n">detach_child</span><span class="p">(</span><span class="n">model</span><span class="p">)</span>

    <span class="c1"># Update the pandas dataframe to reflect the changes</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">generate_df</span><span class="p">()</span></div>


<div class="viewcode-block" id="Protein.save">
<a class="viewcode-back" href="../../neurosnap.html#neurosnap.protein.Protein.save">[docs]</a>
  <span class="k">def</span><span class="w"> </span><span class="nf">save</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">fpath</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="nb">format</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;auto&quot;</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Save the structure as a PDB or mmCIF file.</span>
<span class="sd">    Will overwrite any existing files.</span>

<span class="sd">    Parameters:</span>
<span class="sd">      fpath: File path where you want to save the structure</span>
<span class="sd">      format: File format to save in, either &#39;pdb&#39; or &#39;mmcif&#39;, set to &#39;auto&#39; to infer format from extension.</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="nb">format</span> <span class="o">=</span> <span class="nb">format</span><span class="o">.</span><span class="n">lower</span><span class="p">()</span>
    <span class="c1"># infer format from extension if not provided</span>
    <span class="k">if</span> <span class="nb">format</span> <span class="o">==</span> <span class="s2">&quot;auto&quot;</span><span class="p">:</span>
      <span class="k">if</span> <span class="nb">str</span><span class="p">(</span><span class="n">fpath</span><span class="p">)</span><span class="o">.</span><span class="n">endswith</span><span class="p">(</span><span class="s2">&quot;.pdb&quot;</span><span class="p">):</span>
        <span class="nb">format</span> <span class="o">=</span> <span class="s2">&quot;pdb&quot;</span>
      <span class="k">elif</span> <span class="nb">str</span><span class="p">(</span><span class="n">fpath</span><span class="p">)</span><span class="o">.</span><span class="n">endswith</span><span class="p">(</span><span class="s2">&quot;.mmcif&quot;</span><span class="p">)</span> <span class="ow">or</span> <span class="nb">str</span><span class="p">(</span><span class="n">fpath</span><span class="p">)</span><span class="o">.</span><span class="n">endswith</span><span class="p">(</span><span class="s2">&quot;.cif&quot;</span><span class="p">):</span>
        <span class="nb">format</span> <span class="o">=</span> <span class="s2">&quot;mmcif&quot;</span>
      <span class="k">else</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;Failed to infer format for file. Supported extensions consist of &#39;.pdb&#39; or &#39;.mmcif&#39;.&quot;</span><span class="p">)</span>

    <span class="c1"># save file</span>
    <span class="k">if</span> <span class="nb">format</span> <span class="o">==</span> <span class="s2">&quot;pdb&quot;</span><span class="p">:</span>
      <span class="n">io</span> <span class="o">=</span> <span class="n">PDBIO</span><span class="p">()</span>
      <span class="n">io</span><span class="o">.</span><span class="n">set_structure</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">structure</span><span class="p">)</span>
      <span class="n">io</span><span class="o">.</span><span class="n">save</span><span class="p">(</span><span class="n">fpath</span><span class="p">,</span> <span class="n">preserve_atom_numbering</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
    <span class="k">elif</span> <span class="nb">format</span> <span class="o">==</span> <span class="s2">&quot;mmcif&quot;</span><span class="p">:</span>
      <span class="n">mmcif_io</span> <span class="o">=</span> <span class="n">MMCIFIO</span><span class="p">()</span>
      <span class="n">mmcif_io</span><span class="o">.</span><span class="n">set_structure</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">structure</span><span class="p">)</span>
      <span class="n">mmcif_io</span><span class="o">.</span><span class="n">save</span><span class="p">(</span><span class="n">fpath</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
      <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;Format must be &#39;pdb&#39; or &#39;mmcif&#39;.&quot;</span><span class="p">)</span></div>
</div>



<span class="c1">### FUNCTIONS ###</span>
<div class="viewcode-block" id="getAA">
<a class="viewcode-back" href="../../neurosnap.html#neurosnap.protein.getAA">[docs]</a>
<span class="k">def</span><span class="w"> </span><span class="nf">getAA</span><span class="p">(</span><span class="n">query</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Tuple</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">str</span><span class="p">,</span> <span class="nb">str</span><span class="p">]:</span>
<span class="w">  </span><span class="sd">&quot;&quot;&quot;Efficiently get any amino acid using either their 1 letter code,</span>
<span class="sd">  3 letter abbreviation, or full name. See AAs_FULL_TABLE</span>
<span class="sd">  for a list of all supported amino acids and codes.</span>

<span class="sd">  Parameters:</span>
<span class="sd">    query: Amino acid code, abbreviation, or name</span>

<span class="sd">  Returns:</span>
<span class="sd">    A triple of the form ``(code, abr, name)``.</span>

<span class="sd">    - ``code`` is the amino acid 1 letter abbreviation / code</span>
<span class="sd">    - ``abr`` is the amino acid 3 letter abbreviation / code</span>
<span class="sd">    - ``name`` is the amino acid full name</span>

<span class="sd">  &quot;&quot;&quot;</span>
  <span class="n">query</span> <span class="o">=</span> <span class="n">query</span><span class="o">.</span><span class="n">upper</span><span class="p">()</span>
  <span class="k">try</span><span class="p">:</span>
    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">query</span><span class="p">)</span> <span class="o">==</span> <span class="mi">1</span><span class="p">:</span>
      <span class="k">return</span> <span class="n">query</span><span class="p">,</span> <span class="n">AA_CODE_TO_ABR</span><span class="p">[</span><span class="n">query</span><span class="p">],</span> <span class="n">AA_CODE_TO_NAME</span><span class="p">[</span><span class="n">query</span><span class="p">]</span>
    <span class="k">elif</span> <span class="nb">len</span><span class="p">(</span><span class="n">query</span><span class="p">)</span> <span class="o">==</span> <span class="mi">3</span><span class="p">:</span>
      <span class="k">return</span> <span class="n">AA_ABR_TO_CODE</span><span class="p">[</span><span class="n">query</span><span class="p">],</span> <span class="n">query</span><span class="p">,</span> <span class="n">AA_ABR_TO_NAME</span><span class="p">[</span><span class="n">query</span><span class="p">]</span>
    <span class="k">else</span><span class="p">:</span>
      <span class="k">return</span> <span class="n">AA_NAME_TO_CODE</span><span class="p">[</span><span class="n">query</span><span class="p">],</span> <span class="n">AA_NAME_TO_ABR</span><span class="p">[</span><span class="n">query</span><span class="p">],</span> <span class="n">query</span>
  <span class="k">except</span> <span class="ne">KeyError</span><span class="p">:</span>
    <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Unknown amino acid for </span><span class="si">{</span><span class="n">query</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span></div>



<div class="viewcode-block" id="extract_non_biopolymers">
<a class="viewcode-back" href="../../neurosnap.html#neurosnap.protein.extract_non_biopolymers">[docs]</a>
<span class="k">def</span><span class="w"> </span><span class="nf">extract_non_biopolymers</span><span class="p">(</span><span class="n">pdb_file</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">output_dir</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">min_atoms</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">0</span><span class="p">):</span>
<span class="w">  </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">  Extracts all non-biopolymer molecules (ligands, heteroatoms, etc.)</span>
<span class="sd">  from the specified PDB file and writes them to SDF files.</span>
<span class="sd">  Each molecule is saved as a separate SDF file in the output directory.</span>
<span class="sd">  Automatically adds hydrogens to molecules. Attempts to sanitize</span>
<span class="sd">  the molecule if possible; logs a warning if sanitization fails.</span>

<span class="sd">  Parameters:</span>
<span class="sd">      pdb_file: Path to the input PDB file.</span>
<span class="sd">      output_dir: Directory where the SDF files will be saved. Will overwrite existing directory.</span>
<span class="sd">      min_atoms: Minimum number of atoms a molecule must have to be saved. Molecules with fewer atoms are skipped.</span>
<span class="sd">  &quot;&quot;&quot;</span>

  <span class="k">def</span><span class="w"> </span><span class="nf">is_biopolymer</span><span class="p">(</span><span class="n">molecule</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Determines if a molecule is a biopolymer (protein or nucleotide) based on specific characteristics.</span>
<span class="sd">    Returns True if it is a biopolymer; False otherwise.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Check for peptide bonds or nucleotide backbones</span>
    <span class="c1"># Simplified logic: exclude molecules with standard amino acids or nucleotide bases</span>
    <span class="n">biopolymer_keywords</span> <span class="o">=</span> <span class="p">[</span>
      <span class="s2">&quot;GLY&quot;</span><span class="p">,</span>
      <span class="s2">&quot;ALA&quot;</span><span class="p">,</span>
      <span class="s2">&quot;VAL&quot;</span><span class="p">,</span>
      <span class="s2">&quot;LEU&quot;</span><span class="p">,</span>
      <span class="s2">&quot;ILE&quot;</span><span class="p">,</span>
      <span class="s2">&quot;MET&quot;</span><span class="p">,</span>
      <span class="s2">&quot;PHE&quot;</span><span class="p">,</span>
      <span class="s2">&quot;TYR&quot;</span><span class="p">,</span>
      <span class="s2">&quot;TRP&quot;</span><span class="p">,</span>
      <span class="s2">&quot;SER&quot;</span><span class="p">,</span>
      <span class="s2">&quot;THR&quot;</span><span class="p">,</span>
      <span class="s2">&quot;CYS&quot;</span><span class="p">,</span>
      <span class="s2">&quot;PRO&quot;</span><span class="p">,</span>
      <span class="s2">&quot;ASN&quot;</span><span class="p">,</span>
      <span class="s2">&quot;GLN&quot;</span><span class="p">,</span>
      <span class="s2">&quot;ASP&quot;</span><span class="p">,</span>
      <span class="s2">&quot;GLU&quot;</span><span class="p">,</span>
      <span class="s2">&quot;LYS&quot;</span><span class="p">,</span>
      <span class="s2">&quot;ARG&quot;</span><span class="p">,</span>
      <span class="s2">&quot;HIS&quot;</span><span class="p">,</span>  <span class="c1"># Amino acids</span>
      <span class="s2">&quot;DA&quot;</span><span class="p">,</span>
      <span class="s2">&quot;DT&quot;</span><span class="p">,</span>
      <span class="s2">&quot;DG&quot;</span><span class="p">,</span>
      <span class="s2">&quot;DC&quot;</span><span class="p">,</span>
      <span class="s2">&quot;A&quot;</span><span class="p">,</span>
      <span class="s2">&quot;T&quot;</span><span class="p">,</span>
      <span class="s2">&quot;G&quot;</span><span class="p">,</span>
      <span class="s2">&quot;C&quot;</span><span class="p">,</span>
      <span class="s2">&quot;U&quot;</span><span class="p">,</span>  <span class="c1"># Nucleotide bases</span>
    <span class="p">]</span>
    <span class="k">for</span> <span class="n">atom</span> <span class="ow">in</span> <span class="n">molecule</span><span class="o">.</span><span class="n">GetAtoms</span><span class="p">():</span>
      <span class="n">residue_info</span> <span class="o">=</span> <span class="n">atom</span><span class="o">.</span><span class="n">GetPDBResidueInfo</span><span class="p">()</span>
      <span class="k">if</span> <span class="n">residue_info</span><span class="p">:</span>
        <span class="n">res_name</span> <span class="o">=</span> <span class="n">residue_info</span><span class="o">.</span><span class="n">GetResidueName</span><span class="p">()</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span>
        <span class="k">if</span> <span class="n">res_name</span> <span class="ow">in</span> <span class="n">biopolymer_keywords</span><span class="p">:</span>
          <span class="k">return</span> <span class="kc">True</span>
    <span class="k">return</span> <span class="kc">False</span>

  <span class="c1"># Create output directory if it doesn&#39;t exist</span>
  <span class="k">if</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">exists</span><span class="p">(</span><span class="n">output_dir</span><span class="p">):</span>
    <span class="n">shutil</span><span class="o">.</span><span class="n">rmtree</span><span class="p">(</span><span class="n">output_dir</span><span class="p">)</span>
  <span class="n">os</span><span class="o">.</span><span class="n">makedirs</span><span class="p">(</span><span class="n">output_dir</span><span class="p">)</span>

  <span class="c1"># Read the PDB file</span>
  <span class="n">mol</span> <span class="o">=</span> <span class="n">Chem</span><span class="o">.</span><span class="n">MolFromPDBFile</span><span class="p">(</span><span class="n">pdb_file</span><span class="p">,</span> <span class="n">removeHs</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">sanitize</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
  <span class="k">if</span> <span class="n">mol</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
    <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Failed to read PDB file: </span><span class="si">{</span><span class="n">pdb_file</span><span class="si">}</span><span class="s2">.&quot;</span><span class="p">)</span>

  <span class="c1"># Split the molecule into fragments (separate entities in the PDB)</span>
  <span class="n">fragments</span> <span class="o">=</span> <span class="n">Chem</span><span class="o">.</span><span class="n">GetMolFrags</span><span class="p">(</span><span class="n">mol</span><span class="p">,</span> <span class="n">asMols</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">sanitizeFrags</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
  <span class="n">molecule_count</span> <span class="o">=</span> <span class="mi">1</span>

  <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">frag</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">fragments</span><span class="p">):</span>
    <span class="k">if</span> <span class="n">frag</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
      <span class="n">logger</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Skipping fragment </span><span class="si">{</span><span class="n">i</span><span class="si">}</span><span class="s2"> due to processing failure.&quot;</span><span class="p">)</span>
      <span class="k">continue</span>

    <span class="k">try</span><span class="p">:</span>
      <span class="c1"># Add hydrogens and sanitize molecule</span>
      <span class="n">Chem</span><span class="o">.</span><span class="n">SanitizeMol</span><span class="p">(</span><span class="n">frag</span><span class="p">)</span>
    <span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
      <span class="n">logger</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Failed to sanitize fragment </span><span class="si">{</span><span class="n">i</span><span class="si">}</span><span class="s2">: </span><span class="si">{</span><span class="n">e</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
      <span class="k">continue</span>

    <span class="c1"># Check if the fragment is a biopolymer</span>
    <span class="k">if</span> <span class="n">is_biopolymer</span><span class="p">(</span><span class="n">frag</span><span class="p">):</span>
      <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Skipping biopolymer fragment </span><span class="si">{</span><span class="n">i</span><span class="si">}</span><span class="s2">.&quot;</span><span class="p">)</span>
      <span class="k">continue</span>

    <span class="c1"># Skip small molecules based on atom count</span>
    <span class="k">if</span> <span class="n">frag</span><span class="o">.</span><span class="n">GetNumAtoms</span><span class="p">()</span> <span class="o">&lt;</span> <span class="n">min_atoms</span><span class="p">:</span>
      <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Skipping small molecule fragment </span><span class="si">{</span><span class="n">i</span><span class="si">}</span><span class="s2"> (atom count: </span><span class="si">{</span><span class="n">frag</span><span class="o">.</span><span class="n">GetNumAtoms</span><span class="p">()</span><span class="si">}</span><span class="s2">).&quot;</span><span class="p">)</span>
      <span class="k">continue</span>

    <span class="c1"># Save fragment to SDF</span>
    <span class="n">sdf_file</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">output_dir</span><span class="p">,</span> <span class="sa">f</span><span class="s2">&quot;ligand_</span><span class="si">{</span><span class="n">molecule_count</span><span class="si">}</span><span class="s2">.sdf&quot;</span><span class="p">)</span>
    <span class="n">writer</span> <span class="o">=</span> <span class="n">Chem</span><span class="o">.</span><span class="n">SDWriter</span><span class="p">(</span><span class="n">sdf_file</span><span class="p">)</span>
    <span class="n">writer</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">frag</span><span class="p">)</span>
    <span class="n">writer</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
    <span class="n">molecule_count</span> <span class="o">+=</span> <span class="mi">1</span>

  <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Extracted </span><span class="si">{</span><span class="n">molecule_count</span><span class="o">-</span><span class="mi">1</span><span class="si">}</span><span class="s2"> non-biopolymer molecules to </span><span class="si">{</span><span class="n">output_dir</span><span class="si">}</span><span class="s2">.&quot;</span><span class="p">)</span></div>



<div class="viewcode-block" id="calc_lDDT">
<a class="viewcode-back" href="../../neurosnap.html#neurosnap.protein.calc_lDDT">[docs]</a>
<span class="k">def</span><span class="w"> </span><span class="nf">calc_lDDT</span><span class="p">(</span><span class="n">ref_pdb</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">sample_pdb</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">float</span><span class="p">:</span>
<span class="w">  </span><span class="sd">&quot;&quot;&quot;Calculates the lDDT (Local Distance Difference Test) between two proteins.</span>

<span class="sd">  Parameters:</span>
<span class="sd">    ref_pdb: Filepath for reference protein</span>
<span class="sd">    sample_pdb: Filepath for sample protein</span>

<span class="sd">  Returns:</span>
<span class="sd">    The lDDT score of the two proteins which ranges between 0-1</span>

<span class="sd">  &quot;&quot;&quot;</span>
  <span class="n">ref_L</span><span class="p">,</span> <span class="n">ref_dmap</span><span class="p">,</span> <span class="n">ref_rnames</span> <span class="o">=</span> <span class="n">lDDT</span><span class="o">.</span><span class="n">pdb2dmap</span><span class="p">(</span><span class="n">ref_pdb</span><span class="p">)</span>
  <span class="n">mod_L</span><span class="p">,</span> <span class="n">mod_dmap</span><span class="p">,</span> <span class="n">mod_rnames</span> <span class="o">=</span> <span class="n">lDDT</span><span class="o">.</span><span class="n">pdb2dmap</span><span class="p">(</span><span class="n">sample_pdb</span><span class="p">)</span>
  <span class="k">return</span> <span class="n">lDDT</span><span class="o">.</span><span class="n">get_LDDT</span><span class="p">(</span><span class="n">ref_dmap</span><span class="p">,</span> <span class="n">mod_dmap</span><span class="p">)</span></div>



<div class="viewcode-block" id="foldseek_search">
<a class="viewcode-back" href="../../neurosnap.html#neurosnap.protein.foldseek_search">[docs]</a>
<span class="k">def</span><span class="w"> </span><span class="nf">foldseek_search</span><span class="p">(</span>
  <span class="n">protein</span><span class="p">:</span> <span class="n">Union</span><span class="p">[</span><span class="s2">&quot;Protein&quot;</span><span class="p">,</span> <span class="nb">str</span><span class="p">],</span>
  <span class="n">mode</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;3diaa&quot;</span><span class="p">,</span>
  <span class="n">databases</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
  <span class="n">max_retries</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">10</span><span class="p">,</span>
  <span class="n">retry_interval</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">5</span><span class="p">,</span>
  <span class="n">output_format</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;json&quot;</span><span class="p">,</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Union</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">]:</span>
<span class="w">  </span><span class="sd">&quot;&quot;&quot;Perform a protein structure search using the Foldseek API.</span>

<span class="sd">  Parameters:</span>
<span class="sd">      protein: Either a Protein object or a path to a PDB file.</span>
<span class="sd">      mode: Search mode. Must be on of &quot;3diaa&quot; or &quot;tm-align&quot;.</span>
<span class="sd">      databases: List of databases to search. Defaults to a predefined list if not provided.</span>
<span class="sd">      max_retries: Maximum number of retries to check the job status.</span>
<span class="sd">      retry_interval: Time in seconds between retries for checking job status.</span>
<span class="sd">      output_format: Format of the output, either &quot;json&quot; or &quot;dataframe&quot;.</span>

<span class="sd">  Returns:</span>
<span class="sd">      Search results in the specified format (JSON string or pandas DataFrame).</span>

<span class="sd">  Raises:</span>
<span class="sd">      RuntimeError: If the job fails.</span>
<span class="sd">      TimeoutError: If the job does not complete within the allotted retries.</span>
<span class="sd">      ValueError: If an invalid output_format is specified.</span>

<span class="sd">  &quot;&quot;&quot;</span>

  <span class="n">BASE_URL</span> <span class="o">=</span> <span class="s2">&quot;https://search.foldseek.com/api&quot;</span>

  <span class="c1"># Default databases to search</span>
  <span class="k">if</span> <span class="n">databases</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
    <span class="n">databases</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;afdb50&quot;</span><span class="p">,</span> <span class="s2">&quot;afdb-swissprot&quot;</span><span class="p">,</span> <span class="s2">&quot;afdb-proteome&quot;</span><span class="p">,</span> <span class="s2">&quot;bfmd&quot;</span><span class="p">,</span> <span class="s2">&quot;cath50&quot;</span><span class="p">,</span> <span class="s2">&quot;mgnify_esm30&quot;</span><span class="p">,</span> <span class="s2">&quot;pdb100&quot;</span><span class="p">,</span> <span class="s2">&quot;gmgcl_id&quot;</span><span class="p">,</span> <span class="s2">&quot;bfvd&quot;</span><span class="p">]</span>

  <span class="c1"># Handle file input (Protein object or file path)</span>
  <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">protein</span><span class="p">,</span> <span class="n">Protein</span><span class="p">):</span>
    <span class="k">with</span> <span class="n">tempfile</span><span class="o">.</span><span class="n">NamedTemporaryFile</span><span class="p">(</span><span class="n">delete</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">suffix</span><span class="o">=</span><span class="s2">&quot;.pdb&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">temp_file</span><span class="p">:</span>
      <span class="n">protein</span><span class="o">.</span><span class="n">save</span><span class="p">(</span><span class="n">temp_file</span><span class="o">.</span><span class="n">name</span><span class="p">)</span>
      <span class="n">file_path</span> <span class="o">=</span> <span class="n">temp_file</span><span class="o">.</span><span class="n">name</span>
  <span class="k">else</span><span class="p">:</span>
    <span class="n">file_path</span> <span class="o">=</span> <span class="n">protein</span>

  <span class="c1"># Submit the job to the Foldseek API</span>
  <span class="n">data</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;mode&quot;</span><span class="p">:</span> <span class="n">mode</span><span class="p">,</span> <span class="s2">&quot;database[]&quot;</span><span class="p">:</span> <span class="n">databases</span><span class="p">}</span>
  <span class="k">try</span><span class="p">:</span>
    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">file_path</span><span class="p">,</span> <span class="s2">&quot;rb&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">file</span><span class="p">:</span>
      <span class="n">files</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;q&quot;</span><span class="p">:</span> <span class="n">file</span><span class="p">}</span>
      <span class="n">response</span> <span class="o">=</span> <span class="n">requests</span><span class="o">.</span><span class="n">post</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">BASE_URL</span><span class="si">}</span><span class="s2">/ticket&quot;</span><span class="p">,</span> <span class="n">data</span><span class="o">=</span><span class="n">data</span><span class="p">,</span> <span class="n">files</span><span class="o">=</span><span class="n">files</span><span class="p">)</span>
    <span class="n">response</span><span class="o">.</span><span class="n">raise_for_status</span><span class="p">()</span>
    <span class="n">job_id</span> <span class="o">=</span> <span class="n">response</span><span class="o">.</span><span class="n">json</span><span class="p">()[</span><span class="s2">&quot;id&quot;</span><span class="p">]</span>
  <span class="k">except</span> <span class="n">requests</span><span class="o">.</span><span class="n">RequestException</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
    <span class="k">raise</span> <span class="ne">RuntimeError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Failed to submit job: </span><span class="si">{</span><span class="n">e</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

  <span class="c1"># Poll for job status until complete or max retries are reached</span>
  <span class="k">for</span> <span class="n">attempt</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">max_retries</span><span class="p">):</span>
    <span class="k">try</span><span class="p">:</span>
      <span class="n">status_response</span> <span class="o">=</span> <span class="n">requests</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">BASE_URL</span><span class="si">}</span><span class="s2">/ticket/</span><span class="si">{</span><span class="n">job_id</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
      <span class="n">status_response</span><span class="o">.</span><span class="n">raise_for_status</span><span class="p">()</span>
      <span class="n">status</span> <span class="o">=</span> <span class="n">status_response</span><span class="o">.</span><span class="n">json</span><span class="p">()</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;status&quot;</span><span class="p">,</span> <span class="s2">&quot;ERROR&quot;</span><span class="p">)</span>
    <span class="k">except</span> <span class="n">requests</span><span class="o">.</span><span class="n">RequestException</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
      <span class="k">raise</span> <span class="ne">RuntimeError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Failed to retrieve job status: </span><span class="si">{</span><span class="n">e</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

    <span class="k">if</span> <span class="n">status</span> <span class="o">==</span> <span class="s2">&quot;COMPLETE&quot;</span><span class="p">:</span>
      <span class="k">break</span>
    <span class="k">elif</span> <span class="n">status</span> <span class="o">==</span> <span class="s2">&quot;ERROR&quot;</span><span class="p">:</span>
      <span class="k">raise</span> <span class="ne">RuntimeError</span><span class="p">(</span><span class="s2">&quot;Job failed&quot;</span><span class="p">)</span>

    <span class="n">time</span><span class="o">.</span><span class="n">sleep</span><span class="p">(</span><span class="n">retry_interval</span><span class="p">)</span>
  <span class="k">else</span><span class="p">:</span>
    <span class="k">raise</span> <span class="ne">TimeoutError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Job did not complete within </span><span class="si">{</span><span class="n">max_retries</span><span class="w"> </span><span class="o">*</span><span class="w"> </span><span class="n">retry_interval</span><span class="si">}</span><span class="s2"> seconds&quot;</span><span class="p">)</span>

  <span class="c1"># Retrieve and accumulate results</span>
  <span class="n">results</span> <span class="o">=</span> <span class="p">[]</span>
  <span class="n">entry</span> <span class="o">=</span> <span class="mi">0</span>
  <span class="k">while</span> <span class="kc">True</span><span class="p">:</span>
    <span class="k">try</span><span class="p">:</span>
      <span class="n">result_response</span> <span class="o">=</span> <span class="n">requests</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">BASE_URL</span><span class="si">}</span><span class="s2">/result/</span><span class="si">{</span><span class="n">job_id</span><span class="si">}</span><span class="s2">/</span><span class="si">{</span><span class="n">entry</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
      <span class="n">result_response</span><span class="o">.</span><span class="n">raise_for_status</span><span class="p">()</span>
      <span class="n">result</span> <span class="o">=</span> <span class="n">result_response</span><span class="o">.</span><span class="n">json</span><span class="p">()</span>
    <span class="k">except</span> <span class="n">requests</span><span class="o">.</span><span class="n">RequestException</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
      <span class="k">raise</span> <span class="ne">RuntimeError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Failed to retrieve results: </span><span class="si">{</span><span class="n">e</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

    <span class="k">if</span> <span class="ow">not</span> <span class="n">result</span> <span class="ow">or</span> <span class="nb">all</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">db_result</span><span class="p">[</span><span class="s2">&quot;alignments&quot;</span><span class="p">])</span> <span class="o">==</span> <span class="mi">0</span> <span class="k">for</span> <span class="n">db_result</span> <span class="ow">in</span> <span class="n">result</span><span class="p">[</span><span class="s2">&quot;results&quot;</span><span class="p">]):</span>
      <span class="k">break</span>

    <span class="n">results</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">result</span><span class="p">)</span>
    <span class="n">entry</span> <span class="o">+=</span> <span class="mi">1</span>

  <span class="c1"># Clean up temporary file if it was created</span>
  <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">protein</span><span class="p">,</span> <span class="n">Protein</span><span class="p">):</span>
    <span class="n">os</span><span class="o">.</span><span class="n">remove</span><span class="p">(</span><span class="n">file_path</span><span class="p">)</span>

  <span class="c1"># Return results based on the output format</span>
  <span class="k">if</span> <span class="n">output_format</span> <span class="o">==</span> <span class="s2">&quot;json&quot;</span><span class="p">:</span>
    <span class="k">return</span> <span class="n">json</span><span class="o">.</span><span class="n">dumps</span><span class="p">(</span><span class="n">results</span><span class="p">,</span> <span class="n">indent</span><span class="o">=</span><span class="mi">2</span><span class="p">)</span>
  <span class="k">elif</span> <span class="n">output_format</span> <span class="o">==</span> <span class="s2">&quot;dataframe&quot;</span><span class="p">:</span>
    <span class="n">rows</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">for</span> <span class="n">result</span> <span class="ow">in</span> <span class="n">results</span><span class="p">:</span>
      <span class="k">for</span> <span class="n">db_result</span> <span class="ow">in</span> <span class="n">result</span><span class="p">[</span><span class="s2">&quot;results&quot;</span><span class="p">]:</span>
        <span class="n">alignments</span> <span class="o">=</span> <span class="n">db_result</span><span class="p">[</span><span class="s2">&quot;alignments&quot;</span><span class="p">]</span>
        <span class="k">for</span> <span class="n">alignment</span> <span class="ow">in</span> <span class="n">alignments</span><span class="p">[</span><span class="mi">0</span><span class="p">]:</span>
          <span class="n">rows</span><span class="o">.</span><span class="n">append</span><span class="p">(</span>
            <span class="p">{</span>
              <span class="s2">&quot;target&quot;</span><span class="p">:</span> <span class="n">alignment</span><span class="p">[</span><span class="s2">&quot;target&quot;</span><span class="p">],</span>
              <span class="s2">&quot;db&quot;</span><span class="p">:</span> <span class="n">db_result</span><span class="p">[</span><span class="s2">&quot;db&quot;</span><span class="p">],</span>
              <span class="s2">&quot;seqId&quot;</span><span class="p">:</span> <span class="n">alignment</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;seqId&quot;</span><span class="p">,</span> <span class="s2">&quot;&quot;</span><span class="p">),</span>
              <span class="s2">&quot;alnLength&quot;</span><span class="p">:</span> <span class="n">alignment</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;alnLength&quot;</span><span class="p">,</span> <span class="s2">&quot;&quot;</span><span class="p">),</span>
              <span class="s2">&quot;missmatches&quot;</span><span class="p">:</span> <span class="n">alignment</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;missmatches&quot;</span><span class="p">,</span> <span class="s2">&quot;&quot;</span><span class="p">),</span>
              <span class="s2">&quot;gapsopened&quot;</span><span class="p">:</span> <span class="n">alignment</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;gapsopened&quot;</span><span class="p">,</span> <span class="s2">&quot;&quot;</span><span class="p">),</span>
              <span class="s2">&quot;qStartPos&quot;</span><span class="p">:</span> <span class="n">alignment</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;qStartPos&quot;</span><span class="p">,</span> <span class="s2">&quot;&quot;</span><span class="p">),</span>
              <span class="s2">&quot;qEndPos&quot;</span><span class="p">:</span> <span class="n">alignment</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;qEndPos&quot;</span><span class="p">,</span> <span class="s2">&quot;&quot;</span><span class="p">),</span>
              <span class="s2">&quot;dbStartPos&quot;</span><span class="p">:</span> <span class="n">alignment</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;dbStartPos&quot;</span><span class="p">,</span> <span class="s2">&quot;&quot;</span><span class="p">),</span>
              <span class="s2">&quot;dbEndPos&quot;</span><span class="p">:</span> <span class="n">alignment</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;dbEndPos&quot;</span><span class="p">,</span> <span class="s2">&quot;&quot;</span><span class="p">),</span>
              <span class="s2">&quot;eval&quot;</span><span class="p">:</span> <span class="n">alignment</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;eval&quot;</span><span class="p">,</span> <span class="s2">&quot;&quot;</span><span class="p">),</span>
              <span class="s2">&quot;score&quot;</span><span class="p">:</span> <span class="n">alignment</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;score&quot;</span><span class="p">,</span> <span class="s2">&quot;&quot;</span><span class="p">),</span>
              <span class="s2">&quot;qLen&quot;</span><span class="p">:</span> <span class="n">alignment</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;qLen&quot;</span><span class="p">,</span> <span class="s2">&quot;&quot;</span><span class="p">),</span>
              <span class="s2">&quot;dbLen&quot;</span><span class="p">:</span> <span class="n">alignment</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;dbLen&quot;</span><span class="p">,</span> <span class="s2">&quot;&quot;</span><span class="p">),</span>
              <span class="s2">&quot;seq&quot;</span><span class="p">:</span> <span class="n">alignment</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;tSeq&quot;</span><span class="p">,</span> <span class="s2">&quot;&quot;</span><span class="p">),</span>
            <span class="p">}</span>
          <span class="p">)</span>
    <span class="k">return</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">rows</span><span class="p">)</span>
  <span class="k">else</span><span class="p">:</span>
    <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;Invalid output_format. Choose &#39;json&#39; or &#39;dataframe&#39;.&quot;</span><span class="p">)</span></div>



<div class="viewcode-block" id="run_blast">
<a class="viewcode-back" href="../../neurosnap.html#neurosnap.protein.run_blast">[docs]</a>
<span class="k">def</span><span class="w"> </span><span class="nf">run_blast</span><span class="p">(</span>
  <span class="n">sequence</span><span class="p">:</span> <span class="n">Union</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="s2">&quot;Protein&quot;</span><span class="p">],</span>
  <span class="n">email</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
  <span class="n">matrix</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;BLOSUM62&quot;</span><span class="p">,</span>
  <span class="n">alignments</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">250</span><span class="p">,</span>
  <span class="n">scores</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">250</span><span class="p">,</span>
  <span class="n">evalue</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mf">10.0</span><span class="p">,</span>
  <span class="nb">filter</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span>
  <span class="n">gapalign</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span>
  <span class="n">database</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;uniprotkb_refprotswissprot&quot;</span><span class="p">,</span>
  <span class="n">output_format</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
  <span class="n">output_path</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
  <span class="n">return_df</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Optional</span><span class="p">[</span><span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">]:</span>
<span class="w">  </span><span class="sd">&quot;&quot;&quot;Submits a BLAST job to the EBI NCBI BLAST web service, checks the status periodically, and retrieves the result.</span>
<span class="sd">  The result can be saved either as an XML or FASTA file. Optionally, a DataFrame with alignment details can be returned.</span>

<span class="sd">  Parameters:</span>
<span class="sd">    sequence: The input amino acid sequence as a string or a Protein object.</span>

<span class="sd">      If a Protein object is provided with multiple chains, an error will be raised, and the user will be prompted to provide a single chain sequence using the :obj:`Protein.get_aas` method.</span>
<span class="sd">    email: The email address to use for communication if there is a problem.</span>
<span class="sd">    matrix: The scoring matrix to use, default is ``&quot;BLOSUM62&quot;``.</span>

<span class="sd">      Must be one of::</span>

<span class="sd">      [&quot;BLOSUM45&quot;, &quot;BLOSUM62&quot;, &quot;BLOSUM80&quot;, &quot;PAM30&quot;, &quot;PAM70&quot;].</span>
<span class="sd">    alignments: The number of alignments to display in the result (default is 250). the number alignments must be one of the following::</span>

<span class="sd">      [50, 100, 250, 500, 750, 1000]</span>
<span class="sd">    scores: The number of scores to display in the result, default is ``250``.</span>
<span class="sd">    evalue: The E threshold for alignments (default is 10.0). Must be one of::</span>

<span class="sd">      [0.00001, 0.0001, 0.001, 0.01, 0.1, 1.0, 10.0, 100.0, 1000.0]</span>
<span class="sd">    filter: Whether to filter low complexity regions (default is False).</span>
<span class="sd">    gapalign: Whether to allow gap alignments (default is True).</span>
<span class="sd">    database: The database to search in, default is ``&quot;uniprotkb_refprotswissprot&quot;``.</span>

<span class="sd">      Must be one of::</span>

<span class="sd">      [&quot;uniprotkb_refprotswissprot&quot;, &quot;uniprotkb_pdb&quot;, &quot;uniprotkb&quot;, &quot;afdb&quot;, &quot;uniprotkb_reference_proteomes&quot;, &quot;uniprotkb_swissprot&quot;, &quot;uniref100&quot;, &quot;uniref90&quot;, &quot;uniref50&quot;, &quot;uniparc&quot;]</span>
<span class="sd">    output_format: The format in which to save the result, either ``&quot;xml&quot;`` or ``&quot;fasta&quot;``. If ``None``, which is the default, no file will be saved.</span>
<span class="sd">    output_path: The file path to save the output. This is required if `output_format` is specified.</span>
<span class="sd">    return_df: Whether to return a DataFrame with alignment details, default is ``True``.</span>

<span class="sd">  Returns:</span>
<span class="sd">    A pandas DataFrame with BLAST hit and alignment information, if `return_df` is True.</span>

<span class="sd">    The DataFrame contains the following columns:</span>
<span class="sd">    - &quot;Hit ID&quot;: The identifier of the hit sequence.</span>
<span class="sd">    - &quot;Accession&quot;: The accession number of the hit sequence.</span>
<span class="sd">    - &quot;Description&quot;: The description of the hit sequence.</span>
<span class="sd">    - &quot;Length&quot;: The length of the hit sequence.</span>
<span class="sd">    - &quot;Score&quot;: The score of the alignment.</span>
<span class="sd">    - &quot;Bits&quot;: The bit score of the alignment.</span>
<span class="sd">    - &quot;Expectation&quot;: The E-value of the alignment.</span>
<span class="sd">    - &quot;Identity (%)&quot;: The percentage identity of the alignment.</span>
<span class="sd">    - &quot;Gaps&quot;: The number of gaps in the alignment.</span>
<span class="sd">    - &quot;Query Sequence&quot;: The query sequence in the alignment.</span>
<span class="sd">    - &quot;Match Sequence&quot;: The matched sequence in the alignment.</span>

<span class="sd">  Raises:</span>
<span class="sd">    AssertionError: If ``sequence`` is provided as a Protein object with multiple chains.</span>
<span class="sd">  &quot;&quot;&quot;</span>

  <span class="k">def</span><span class="w"> </span><span class="nf">parse_xml_to_fasta_and_dataframe</span><span class="p">(</span><span class="n">xml_content</span><span class="p">,</span> <span class="n">job_id</span><span class="p">,</span> <span class="n">output_format</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">output_path</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">return_df</span><span class="o">=</span><span class="kc">True</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Parses the XML content, saves it as a FASTA file, or returns a DataFrame if requested.&quot;&quot;&quot;</span>
    <span class="n">root</span> <span class="o">=</span> <span class="n">ET</span><span class="o">.</span><span class="n">fromstring</span><span class="p">(</span><span class="n">xml_content</span><span class="p">)</span>
    <span class="n">hits</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">fasta_content</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>

    <span class="k">for</span> <span class="n">hit</span> <span class="ow">in</span> <span class="n">root</span><span class="o">.</span><span class="n">findall</span><span class="p">(</span><span class="s2">&quot;.//{http://www.ebi.ac.uk/schema}hit&quot;</span><span class="p">):</span>
      <span class="n">hit_id</span> <span class="o">=</span> <span class="n">hit</span><span class="o">.</span><span class="n">attrib</span><span class="p">[</span><span class="s2">&quot;id&quot;</span><span class="p">]</span>
      <span class="n">hit_ac</span> <span class="o">=</span> <span class="n">hit</span><span class="o">.</span><span class="n">attrib</span><span class="p">[</span><span class="s2">&quot;ac&quot;</span><span class="p">]</span>
      <span class="n">hit_description</span> <span class="o">=</span> <span class="n">hit</span><span class="o">.</span><span class="n">attrib</span><span class="p">[</span><span class="s2">&quot;description&quot;</span><span class="p">]</span>
      <span class="n">hit_length</span> <span class="o">=</span> <span class="n">hit</span><span class="o">.</span><span class="n">attrib</span><span class="p">[</span><span class="s2">&quot;length&quot;</span><span class="p">]</span>

      <span class="k">for</span> <span class="n">alignment</span> <span class="ow">in</span> <span class="n">hit</span><span class="o">.</span><span class="n">findall</span><span class="p">(</span><span class="s2">&quot;.//{http://www.ebi.ac.uk/schema}alignment&quot;</span><span class="p">):</span>
        <span class="n">score</span> <span class="o">=</span> <span class="n">alignment</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="s2">&quot;{http://www.ebi.ac.uk/schema}score&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">text</span>
        <span class="n">bits</span> <span class="o">=</span> <span class="n">alignment</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="s2">&quot;{http://www.ebi.ac.uk/schema}bits&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">text</span>
        <span class="n">expectation</span> <span class="o">=</span> <span class="n">alignment</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="s2">&quot;{http://www.ebi.ac.uk/schema}expectation&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">text</span>
        <span class="n">identity</span> <span class="o">=</span> <span class="n">alignment</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="s2">&quot;{http://www.ebi.ac.uk/schema}identity&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">text</span>
        <span class="n">gaps</span> <span class="o">=</span> <span class="n">alignment</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="s2">&quot;{http://www.ebi.ac.uk/schema}gaps&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">text</span>
        <span class="n">query_seq</span> <span class="o">=</span> <span class="n">alignment</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="s2">&quot;{http://www.ebi.ac.uk/schema}querySeq&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">text</span>
        <span class="n">match_seq</span> <span class="o">=</span> <span class="n">alignment</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="s2">&quot;{http://www.ebi.ac.uk/schema}matchSeq&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">text</span>

        <span class="n">fasta_content</span> <span class="o">+=</span> <span class="p">(</span>
          <span class="sa">f</span><span class="s2">&quot;&gt;</span><span class="si">{</span><span class="n">hit_id</span><span class="si">}</span><span class="s2"> | Accession: </span><span class="si">{</span><span class="n">hit_ac</span><span class="si">}</span><span class="s2"> | Description: </span><span class="si">{</span><span class="n">hit_description</span><span class="si">}</span><span class="s2"> | &quot;</span>
          <span class="sa">f</span><span class="s2">&quot;Length: </span><span class="si">{</span><span class="n">hit_length</span><span class="si">}</span><span class="s2"> | Score: </span><span class="si">{</span><span class="n">score</span><span class="si">}</span><span class="s2"> | Bits: </span><span class="si">{</span><span class="n">bits</span><span class="si">}</span><span class="s2"> | &quot;</span>
          <span class="sa">f</span><span class="s2">&quot;Expectation: </span><span class="si">{</span><span class="n">expectation</span><span class="si">}</span><span class="s2"> | Identity: </span><span class="si">{</span><span class="n">identity</span><span class="si">}</span><span class="s2">% | Gaps: </span><span class="si">{</span><span class="n">gaps</span><span class="si">}</span><span class="se">\n</span><span class="s2">&quot;</span>
          <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">match_seq</span><span class="si">}</span><span class="se">\n\n</span><span class="s2">&quot;</span>
        <span class="p">)</span>

        <span class="n">hits</span><span class="o">.</span><span class="n">append</span><span class="p">(</span>
          <span class="p">{</span>
            <span class="s2">&quot;Hit ID&quot;</span><span class="p">:</span> <span class="n">hit_id</span><span class="p">,</span>
            <span class="s2">&quot;Accession&quot;</span><span class="p">:</span> <span class="n">hit_ac</span><span class="p">,</span>
            <span class="s2">&quot;Description&quot;</span><span class="p">:</span> <span class="n">hit_description</span><span class="p">,</span>
            <span class="s2">&quot;Length&quot;</span><span class="p">:</span> <span class="n">hit_length</span><span class="p">,</span>
            <span class="s2">&quot;Score&quot;</span><span class="p">:</span> <span class="n">score</span><span class="p">,</span>
            <span class="s2">&quot;Bits&quot;</span><span class="p">:</span> <span class="n">bits</span><span class="p">,</span>
            <span class="s2">&quot;Expectation&quot;</span><span class="p">:</span> <span class="n">expectation</span><span class="p">,</span>
            <span class="s2">&quot;Identity (%)&quot;</span><span class="p">:</span> <span class="n">identity</span><span class="p">,</span>
            <span class="s2">&quot;Gaps&quot;</span><span class="p">:</span> <span class="n">gaps</span><span class="p">,</span>
            <span class="s2">&quot;Query Sequence&quot;</span><span class="p">:</span> <span class="n">query_seq</span><span class="p">,</span>
            <span class="s2">&quot;Match Sequence&quot;</span><span class="p">:</span> <span class="n">match_seq</span><span class="p">,</span>
          <span class="p">}</span>
        <span class="p">)</span>

    <span class="c1"># Step 4: Save or return results</span>
    <span class="k">if</span> <span class="n">output_format</span> <span class="o">==</span> <span class="s2">&quot;fasta&quot;</span> <span class="ow">and</span> <span class="n">output_path</span><span class="p">:</span>
      <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">output_path</span><span class="p">,</span> <span class="s2">&quot;w&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">fasta_file</span><span class="p">:</span>
        <span class="n">fasta_file</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">fasta_content</span><span class="p">)</span>
      <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;FASTA result saved as </span><span class="si">{</span><span class="n">output_path</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

    <span class="k">if</span> <span class="n">return_df</span><span class="p">:</span>
      <span class="n">df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">hits</span><span class="p">)</span>
      <span class="k">return</span> <span class="n">df</span>

  <span class="n">valid_databases</span> <span class="o">=</span> <span class="p">[</span>
    <span class="s2">&quot;uniprotkb_refprotswissprot&quot;</span><span class="p">,</span>
    <span class="s2">&quot;uniprotkb_pdb&quot;</span><span class="p">,</span>
    <span class="s2">&quot;uniprotkb&quot;</span><span class="p">,</span>
    <span class="s2">&quot;afdb&quot;</span><span class="p">,</span>
    <span class="s2">&quot;uniprotkb_reference_proteomes&quot;</span><span class="p">,</span>
    <span class="s2">&quot;uniprotkb_swissprot&quot;</span><span class="p">,</span>
    <span class="s2">&quot;uniref100&quot;</span><span class="p">,</span>
    <span class="s2">&quot;uniref90&quot;</span><span class="p">,</span>
    <span class="s2">&quot;uniref50&quot;</span><span class="p">,</span>
    <span class="s2">&quot;uniparc&quot;</span><span class="p">,</span>
  <span class="p">]</span>
  <span class="k">if</span> <span class="n">database</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">valid_databases</span><span class="p">:</span>
    <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Database must be one of the following </span><span class="si">{</span><span class="n">valid_databases</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

  <span class="n">valid_matrices</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;BLOSUM45&quot;</span><span class="p">,</span> <span class="s2">&quot;BLOSUM62&quot;</span><span class="p">,</span> <span class="s2">&quot;BLOSUM80&quot;</span><span class="p">,</span> <span class="s2">&quot;PAM30&quot;</span><span class="p">,</span> <span class="s2">&quot;PAM70&quot;</span><span class="p">]</span>
  <span class="k">if</span> <span class="n">matrix</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">valid_matrices</span><span class="p">:</span>
    <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Matrix must be one of the following </span><span class="si">{</span><span class="n">valid_matrices</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

  <span class="n">valid_evalues</span> <span class="o">=</span> <span class="p">[</span><span class="mf">0.00001</span><span class="p">,</span> <span class="mf">0.0001</span><span class="p">,</span> <span class="mf">0.001</span><span class="p">,</span> <span class="mf">0.01</span><span class="p">,</span> <span class="mf">0.1</span><span class="p">,</span> <span class="mf">1.0</span><span class="p">,</span> <span class="mf">10.0</span><span class="p">,</span> <span class="mf">100.0</span><span class="p">,</span> <span class="mf">1000.0</span><span class="p">]</span>
  <span class="k">if</span> <span class="n">evalue</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">valid_evalues</span><span class="p">:</span>
    <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;E-threshold must be one of the following </span><span class="si">{</span><span class="n">valid_evalues</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

  <span class="c1"># the api is very delicate, we need specific parameters</span>
  <span class="k">if</span> <span class="n">evalue</span> <span class="o">&gt;</span> <span class="mi">1</span><span class="p">:</span>
    <span class="n">evalue</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">evalue</span><span class="p">)</span>

  <span class="n">valid_alignments</span> <span class="o">=</span> <span class="p">[</span><span class="mi">50</span><span class="p">,</span> <span class="mi">100</span><span class="p">,</span> <span class="mi">250</span><span class="p">,</span> <span class="mi">500</span><span class="p">,</span> <span class="mi">750</span><span class="p">,</span> <span class="mi">1000</span><span class="p">]</span>
  <span class="k">if</span> <span class="n">alignments</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">valid_alignments</span><span class="p">:</span>
    <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Alignments must be one of the following </span><span class="si">{</span><span class="n">valid_alignments</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

  <span class="n">valid_output_formats</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;xml&quot;</span><span class="p">,</span> <span class="s2">&quot;fasta&quot;</span><span class="p">,</span> <span class="kc">None</span><span class="p">]</span>
  <span class="k">if</span> <span class="n">output_format</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">valid_output_formats</span><span class="p">:</span>
    <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Output format must be one of the following </span><span class="si">{</span><span class="n">valid_output_formats</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

  <span class="c1"># Handle Protein object input</span>
  <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">sequence</span><span class="p">,</span> <span class="n">Protein</span><span class="p">):</span>
    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">sequence</span><span class="o">.</span><span class="n">chains</span><span class="p">())</span> <span class="o">&gt;</span> <span class="mi">1</span><span class="p">:</span>
      <span class="k">raise</span> <span class="ne">AssertionError</span><span class="p">(</span><span class="s2">&quot;The protein has multiple chains. Use &#39;.get_aas(model, chain)&#39; to obtain the sequence for a specific chain.&quot;</span><span class="p">)</span>

    <span class="n">model</span> <span class="o">=</span> <span class="n">sequence</span><span class="o">.</span><span class="n">models</span><span class="p">()[</span><span class="mi">0</span><span class="p">]</span>
    <span class="n">chain</span> <span class="o">=</span> <span class="n">sequence</span><span class="o">.</span><span class="n">chains</span><span class="p">()[</span><span class="mi">0</span><span class="p">]</span>
    <span class="n">sequence</span> <span class="o">=</span> <span class="n">sequence</span><span class="o">.</span><span class="n">get_aas</span><span class="p">(</span><span class="n">model</span><span class="p">,</span> <span class="n">chain</span><span class="p">)</span>

  <span class="c1"># Step 1: Submit the BLAST job</span>
  <span class="n">url</span> <span class="o">=</span> <span class="s2">&quot;https://www.ebi.ac.uk/Tools/services/rest/ncbiblast/run&quot;</span>

  <span class="n">multipart_data</span> <span class="o">=</span> <span class="n">MultipartEncoder</span><span class="p">(</span>
    <span class="n">fields</span><span class="o">=</span><span class="p">{</span>
      <span class="s2">&quot;email&quot;</span><span class="p">:</span> <span class="n">email</span><span class="p">,</span>
      <span class="s2">&quot;program&quot;</span><span class="p">:</span> <span class="s2">&quot;blastp&quot;</span><span class="p">,</span>
      <span class="s2">&quot;matrix&quot;</span><span class="p">:</span> <span class="n">matrix</span><span class="p">,</span>
      <span class="s2">&quot;alignments&quot;</span><span class="p">:</span> <span class="nb">str</span><span class="p">(</span><span class="n">alignments</span><span class="p">),</span>
      <span class="s2">&quot;scores&quot;</span><span class="p">:</span> <span class="nb">str</span><span class="p">(</span><span class="n">scores</span><span class="p">),</span>
      <span class="s2">&quot;exp&quot;</span><span class="p">:</span> <span class="nb">str</span><span class="p">(</span><span class="n">evalue</span><span class="p">),</span>
      <span class="s2">&quot;filter&quot;</span><span class="p">:</span> <span class="s2">&quot;T&quot;</span> <span class="k">if</span> <span class="nb">filter</span> <span class="k">else</span> <span class="s2">&quot;F&quot;</span><span class="p">,</span>
      <span class="s2">&quot;gapalign&quot;</span><span class="p">:</span> <span class="nb">str</span><span class="p">(</span><span class="n">gapalign</span><span class="p">)</span><span class="o">.</span><span class="n">lower</span><span class="p">(),</span>
      <span class="s2">&quot;stype&quot;</span><span class="p">:</span> <span class="s2">&quot;protein&quot;</span><span class="p">,</span>
      <span class="s2">&quot;sequence&quot;</span><span class="p">:</span> <span class="n">sequence</span><span class="p">,</span>
      <span class="s2">&quot;database&quot;</span><span class="p">:</span> <span class="n">database</span><span class="p">,</span>
    <span class="p">}</span>
  <span class="p">)</span>

  <span class="n">headers</span> <span class="o">=</span> <span class="p">{</span>
    <span class="s2">&quot;User-Agent&quot;</span><span class="p">:</span> <span class="n">USER_AGENT</span><span class="p">,</span>
    <span class="s2">&quot;Accept&quot;</span><span class="p">:</span> <span class="s2">&quot;text/plain,application/json&quot;</span><span class="p">,</span>
    <span class="s2">&quot;Accept-Language&quot;</span><span class="p">:</span> <span class="s2">&quot;en-US,en;q=0.5&quot;</span><span class="p">,</span>
    <span class="s2">&quot;Content-Type&quot;</span><span class="p">:</span> <span class="n">multipart_data</span><span class="o">.</span><span class="n">content_type</span><span class="p">,</span>
  <span class="p">}</span>

  <span class="n">response</span> <span class="o">=</span> <span class="n">requests</span><span class="o">.</span><span class="n">post</span><span class="p">(</span><span class="n">url</span><span class="p">,</span> <span class="n">headers</span><span class="o">=</span><span class="n">headers</span><span class="p">,</span> <span class="n">data</span><span class="o">=</span><span class="n">multipart_data</span><span class="p">)</span>

  <span class="k">if</span> <span class="n">response</span><span class="o">.</span><span class="n">status_code</span> <span class="o">==</span> <span class="mi">200</span><span class="p">:</span>
    <span class="n">job_id</span> <span class="o">=</span> <span class="n">response</span><span class="o">.</span><span class="n">text</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span>
    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Job submitted successfully. Job ID: </span><span class="si">{</span><span class="n">job_id</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
  <span class="k">else</span><span class="p">:</span>
    <span class="n">response</span><span class="o">.</span><span class="n">raise_for_status</span><span class="p">()</span>

  <span class="c1"># Step 2: Check job status</span>
  <span class="n">status_url</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;https://www.ebi.ac.uk/Tools/services/rest/ncbiblast/status/</span><span class="si">{</span><span class="n">job_id</span><span class="si">}</span><span class="s2">&quot;</span>

  <span class="k">while</span> <span class="kc">True</span><span class="p">:</span>
    <span class="n">status_response</span> <span class="o">=</span> <span class="n">requests</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">status_url</span><span class="p">)</span>
    <span class="n">status</span> <span class="o">=</span> <span class="n">status_response</span><span class="o">.</span><span class="n">text</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span>

    <span class="k">if</span> <span class="n">status_response</span><span class="o">.</span><span class="n">status_code</span> <span class="o">==</span> <span class="mi">200</span><span class="p">:</span>
      <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Job status: </span><span class="si">{</span><span class="n">status</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
      <span class="k">if</span> <span class="n">status</span> <span class="o">==</span> <span class="s2">&quot;FINISHED&quot;</span><span class="p">:</span>
        <span class="k">break</span>
      <span class="k">elif</span> <span class="n">status</span> <span class="ow">in</span> <span class="p">[</span><span class="s2">&quot;RUNNING&quot;</span><span class="p">,</span> <span class="s2">&quot;PENDING&quot;</span><span class="p">,</span> <span class="s2">&quot;QUEUED&quot;</span><span class="p">]:</span>
        <span class="n">time</span><span class="o">.</span><span class="n">sleep</span><span class="p">(</span><span class="mi">20</span><span class="p">)</span>
      <span class="k">else</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">Exception</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Job failed with status: </span><span class="si">{</span><span class="n">status</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
      <span class="n">status_response</span><span class="o">.</span><span class="n">raise_for_status</span><span class="p">()</span>

  <span class="c1"># Step 3: Retrieve XML result</span>
  <span class="n">xml_url</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;https://www.ebi.ac.uk/Tools/services/rest/ncbiblast/result/</span><span class="si">{</span><span class="n">job_id</span><span class="si">}</span><span class="s2">/xml&quot;</span>
  <span class="n">xml_response</span> <span class="o">=</span> <span class="n">requests</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">xml_url</span><span class="p">)</span>

  <span class="k">if</span> <span class="n">xml_response</span><span class="o">.</span><span class="n">status_code</span> <span class="o">==</span> <span class="mi">200</span><span class="p">:</span>
    <span class="n">xml_content</span> <span class="o">=</span> <span class="n">xml_response</span><span class="o">.</span><span class="n">text</span>
    <span class="c1"># Save XML if output format is XML</span>
    <span class="k">if</span> <span class="n">output_format</span> <span class="o">==</span> <span class="s2">&quot;xml&quot;</span> <span class="ow">and</span> <span class="n">output_path</span><span class="p">:</span>
      <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">output_path</span><span class="p">,</span> <span class="s2">&quot;w&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">xml_file</span><span class="p">:</span>
        <span class="n">xml_file</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">xml_content</span><span class="p">)</span>
      <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;XML result saved as </span><span class="si">{</span><span class="n">output_path</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="k">elif</span> <span class="n">output_format</span> <span class="o">==</span> <span class="s2">&quot;fasta&quot;</span> <span class="ow">and</span> <span class="n">output_path</span><span class="p">:</span>
      <span class="k">return</span> <span class="n">parse_xml_to_fasta_and_dataframe</span><span class="p">(</span><span class="n">xml_content</span><span class="p">,</span> <span class="n">job_id</span><span class="p">,</span> <span class="n">output_format</span><span class="p">,</span> <span class="n">output_path</span><span class="p">,</span> <span class="n">return_df</span><span class="p">)</span>
    <span class="k">elif</span> <span class="n">return_df</span><span class="p">:</span>
      <span class="k">return</span> <span class="n">parse_xml_to_fasta_and_dataframe</span><span class="p">(</span><span class="n">xml_content</span><span class="p">,</span> <span class="n">job_id</span><span class="p">,</span> <span class="n">output_format</span><span class="p">,</span> <span class="n">output_path</span><span class="p">,</span> <span class="n">return_df</span><span class="p">)</span>
  <span class="k">else</span><span class="p">:</span>
    <span class="n">xml_response</span><span class="o">.</span><span class="n">raise_for_status</span><span class="p">()</span></div>



<div class="viewcode-block" id="plot_pseudo_3D">
<a class="viewcode-back" href="../../neurosnap.html#neurosnap.protein.plot_pseudo_3D">[docs]</a>
<span class="k">def</span><span class="w"> </span><span class="nf">plot_pseudo_3D</span><span class="p">(</span>
  <span class="n">xyz</span><span class="p">:</span> <span class="n">Union</span><span class="p">[</span><span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">],</span>
  <span class="n">c</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
  <span class="n">ax</span><span class="p">:</span> <span class="n">matplotlib</span><span class="o">.</span><span class="n">axes</span><span class="o">.</span><span class="n">Axes</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
  <span class="n">chainbreak</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">5</span><span class="p">,</span>
  <span class="n">Ls</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">List</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
  <span class="n">cmap</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;gist_rainbow&quot;</span><span class="p">,</span>
  <span class="n">line_w</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mf">2.0</span><span class="p">,</span>
  <span class="n">cmin</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">float</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
  <span class="n">cmax</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">float</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
  <span class="n">zmin</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">float</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
  <span class="n">zmax</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">float</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
  <span class="n">shadow</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mf">0.95</span><span class="p">,</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">matplotlib</span><span class="o">.</span><span class="n">collections</span><span class="o">.</span><span class="n">LineCollection</span><span class="p">:</span>
<span class="w">  </span><span class="sd">&quot;&quot;&quot;Plot the famous Pseudo 3D projection of a protein.</span>

<span class="sd">  Algorithm originally written By Dr. Sergey Ovchinnikov.</span>
<span class="sd">  Adapted from https://github.com/sokrypton/ColabDesign/blob/16e03c23f2a30a3dcb1775ac25e107424f9f7352/colabdesign/shared/plot.py</span>

<span class="sd">  Parameters:</span>
<span class="sd">    xyz: XYZ coordinates of the protein</span>
<span class="sd">    c: 1D array of all the values to use to color the protein, defaults to residue index</span>
<span class="sd">    ax: Matplotlib axes object to add the figure to</span>
<span class="sd">    chainbreak: Minimum distance in angstroms between chains / segments before being considered a chain break (int)</span>
<span class="sd">    Ls: Allows handling multiple chains or segments by providing the lengths of each chain, ensuring that chains are visualized separately without unwanted connections</span>
<span class="sd">    cmap: Matplotlib color map to use for coloring the protein</span>
<span class="sd">    line_w: Line width</span>
<span class="sd">    cmin: Minimum value for coloring, automatically calculated if None</span>
<span class="sd">    cmax: Maximum value for coloring, automatically calculated if None</span>
<span class="sd">    zmin: Minimum z coordinate values, automatically calculated if None</span>
<span class="sd">    zmax: Maximum z coordinate values, automatically calculated if None</span>
<span class="sd">    shadow: Shadow intensity between 0 and 1 inclusive, lower numbers mean darker more intense shadows</span>

<span class="sd">  Returns:</span>
<span class="sd">    LineCollection object of what&#39;s been drawn</span>

<span class="sd">  &quot;&quot;&quot;</span>

  <span class="k">def</span><span class="w"> </span><span class="nf">rescale</span><span class="p">(</span><span class="n">a</span><span class="p">,</span> <span class="n">amin</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">amax</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
    <span class="n">a</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">copy</span><span class="p">(</span><span class="n">a</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">amin</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
      <span class="n">amin</span> <span class="o">=</span> <span class="n">a</span><span class="o">.</span><span class="n">min</span><span class="p">()</span>
    <span class="k">if</span> <span class="n">amax</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
      <span class="n">amax</span> <span class="o">=</span> <span class="n">a</span><span class="o">.</span><span class="n">max</span><span class="p">()</span>
    <span class="n">a</span><span class="p">[</span><span class="n">a</span> <span class="o">&lt;</span> <span class="n">amin</span><span class="p">]</span> <span class="o">=</span> <span class="n">amin</span>
    <span class="n">a</span><span class="p">[</span><span class="n">a</span> <span class="o">&gt;</span> <span class="n">amax</span><span class="p">]</span> <span class="o">=</span> <span class="n">amax</span>
    <span class="k">return</span> <span class="p">(</span><span class="n">a</span> <span class="o">-</span> <span class="n">amin</span><span class="p">)</span> <span class="o">/</span> <span class="p">(</span><span class="n">amax</span> <span class="o">-</span> <span class="n">amin</span><span class="p">)</span>

  <span class="c1"># clip color values and produce warning if necesarry</span>
  <span class="k">if</span> <span class="n">c</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="ow">and</span> <span class="n">cmin</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="ow">and</span> <span class="n">cmax</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
    <span class="k">if</span> <span class="n">np</span><span class="o">.</span><span class="n">any</span><span class="p">(</span><span class="n">c</span> <span class="o">&lt;</span> <span class="n">cmin</span><span class="p">):</span>
      <span class="n">logger</span><span class="o">.</span><span class="n">warn</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;The provided c colors array contains values that are less than cmin (</span><span class="si">{</span><span class="n">cmin</span><span class="si">}</span><span class="s2">). Out of range values will be clipped into range.&quot;</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">np</span><span class="o">.</span><span class="n">any</span><span class="p">(</span><span class="n">c</span> <span class="o">&gt;</span> <span class="n">cmax</span><span class="p">):</span>
      <span class="n">logger</span><span class="o">.</span><span class="n">warn</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;The provided c colors array contains values that are greater than cmax (</span><span class="si">{</span><span class="n">cmax</span><span class="si">}</span><span class="s2">). Out of range values will be clipped into range.&quot;</span><span class="p">)</span>
    <span class="n">c</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">clip</span><span class="p">(</span><span class="n">c</span><span class="p">,</span> <span class="n">a_min</span><span class="o">=</span><span class="n">cmin</span><span class="p">,</span> <span class="n">a_max</span><span class="o">=</span><span class="n">cmax</span><span class="p">)</span>

  <span class="c1"># make segments and colors for each segment</span>
  <span class="n">xyz</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">asarray</span><span class="p">(</span><span class="n">xyz</span><span class="p">)</span>
  <span class="k">if</span> <span class="n">Ls</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
    <span class="n">seg</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">concatenate</span><span class="p">([</span><span class="n">xyz</span><span class="p">[:,</span> <span class="kc">None</span><span class="p">],</span> <span class="n">np</span><span class="o">.</span><span class="n">roll</span><span class="p">(</span><span class="n">xyz</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">0</span><span class="p">)[:,</span> <span class="kc">None</span><span class="p">]],</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
    <span class="n">c_seg</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">seg</span><span class="p">))[::</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span> <span class="k">if</span> <span class="n">c</span> <span class="ow">is</span> <span class="kc">None</span> <span class="k">else</span> <span class="p">(</span><span class="n">c</span> <span class="o">+</span> <span class="n">np</span><span class="o">.</span><span class="n">roll</span><span class="p">(</span><span class="n">c</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">0</span><span class="p">))</span> <span class="o">/</span> <span class="mi">2</span>
  <span class="k">else</span><span class="p">:</span>
    <span class="n">Ln</span> <span class="o">=</span> <span class="mi">0</span>
    <span class="n">seg</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">c_seg</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">for</span> <span class="n">L</span> <span class="ow">in</span> <span class="n">Ls</span><span class="p">:</span>
      <span class="n">sub_xyz</span> <span class="o">=</span> <span class="n">xyz</span><span class="p">[</span><span class="n">Ln</span> <span class="p">:</span> <span class="n">Ln</span> <span class="o">+</span> <span class="n">L</span><span class="p">]</span>
      <span class="n">seg</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">concatenate</span><span class="p">([</span><span class="n">sub_xyz</span><span class="p">[:,</span> <span class="kc">None</span><span class="p">],</span> <span class="n">np</span><span class="o">.</span><span class="n">roll</span><span class="p">(</span><span class="n">sub_xyz</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">0</span><span class="p">)[:,</span> <span class="kc">None</span><span class="p">]],</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">))</span>
      <span class="k">if</span> <span class="n">c</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">sub_c</span> <span class="o">=</span> <span class="n">c</span><span class="p">[</span><span class="n">Ln</span> <span class="p">:</span> <span class="n">Ln</span> <span class="o">+</span> <span class="n">L</span><span class="p">]</span>
        <span class="n">c_seg</span><span class="o">.</span><span class="n">append</span><span class="p">((</span><span class="n">sub_c</span> <span class="o">+</span> <span class="n">np</span><span class="o">.</span><span class="n">roll</span><span class="p">(</span><span class="n">sub_c</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">0</span><span class="p">))</span> <span class="o">/</span> <span class="mi">2</span><span class="p">)</span>
      <span class="n">Ln</span> <span class="o">+=</span> <span class="n">L</span>
    <span class="n">seg</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">concatenate</span><span class="p">(</span><span class="n">seg</span><span class="p">,</span> <span class="mi">0</span><span class="p">)</span>
    <span class="n">c_seg</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">seg</span><span class="p">))[::</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span> <span class="k">if</span> <span class="n">c</span> <span class="ow">is</span> <span class="kc">None</span> <span class="k">else</span> <span class="n">np</span><span class="o">.</span><span class="n">concatenate</span><span class="p">(</span><span class="n">c_seg</span><span class="p">,</span> <span class="mi">0</span><span class="p">)</span>

  <span class="c1"># set colors</span>
  <span class="n">c_seg</span> <span class="o">=</span> <span class="n">rescale</span><span class="p">(</span><span class="n">c_seg</span><span class="p">,</span> <span class="n">cmin</span><span class="p">,</span> <span class="n">cmax</span><span class="p">)</span>
  <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">cmap</span><span class="p">,</span> <span class="nb">str</span><span class="p">):</span>
    <span class="k">if</span> <span class="n">cmap</span> <span class="o">==</span> <span class="s2">&quot;gist_rainbow&quot;</span><span class="p">:</span>
      <span class="n">c_seg</span> <span class="o">*=</span> <span class="mf">0.75</span>
    <span class="n">colors</span> <span class="o">=</span> <span class="n">matplotlib</span><span class="o">.</span><span class="n">colormaps</span><span class="p">[</span><span class="n">cmap</span><span class="p">](</span><span class="n">c_seg</span><span class="p">)</span>
  <span class="k">else</span><span class="p">:</span>
    <span class="n">colors</span> <span class="o">=</span> <span class="n">cmap</span><span class="p">(</span><span class="n">c_seg</span><span class="p">)</span>

  <span class="c1"># remove segments that aren&#39;t connected</span>
  <span class="n">seg_len</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">sqrt</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">square</span><span class="p">(</span><span class="n">seg</span><span class="p">[:,</span> <span class="mi">0</span><span class="p">]</span> <span class="o">-</span> <span class="n">seg</span><span class="p">[:,</span> <span class="mi">1</span><span class="p">])</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="o">-</span><span class="mi">1</span><span class="p">))</span>
  <span class="k">if</span> <span class="n">chainbreak</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
    <span class="n">idx</span> <span class="o">=</span> <span class="n">seg_len</span> <span class="o">&lt;</span> <span class="n">chainbreak</span>
    <span class="n">seg</span> <span class="o">=</span> <span class="n">seg</span><span class="p">[</span><span class="n">idx</span><span class="p">]</span>
    <span class="n">seg_len</span> <span class="o">=</span> <span class="n">seg_len</span><span class="p">[</span><span class="n">idx</span><span class="p">]</span>
    <span class="n">colors</span> <span class="o">=</span> <span class="n">colors</span><span class="p">[</span><span class="n">idx</span><span class="p">]</span>

  <span class="n">seg_mid</span> <span class="o">=</span> <span class="n">seg</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="mi">1</span><span class="p">)</span>
  <span class="n">seg_xy</span> <span class="o">=</span> <span class="n">seg</span><span class="p">[</span><span class="o">...</span><span class="p">,</span> <span class="p">:</span><span class="mi">2</span><span class="p">]</span>
  <span class="n">seg_z</span> <span class="o">=</span> <span class="n">seg</span><span class="p">[</span><span class="o">...</span><span class="p">,</span> <span class="mi">2</span><span class="p">]</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="o">-</span><span class="mi">1</span><span class="p">)</span>
  <span class="n">order</span> <span class="o">=</span> <span class="n">seg_z</span><span class="o">.</span><span class="n">argsort</span><span class="p">()</span>

  <span class="c1"># add shade/tint based on z-dimension</span>
  <span class="n">z</span> <span class="o">=</span> <span class="n">rescale</span><span class="p">(</span><span class="n">seg_z</span><span class="p">,</span> <span class="n">zmin</span><span class="p">,</span> <span class="n">zmax</span><span class="p">)[:,</span> <span class="kc">None</span><span class="p">]</span>

  <span class="c1"># add shadow (make lines darker if they are behind other lines)</span>
  <span class="n">seg_len_cutoff</span> <span class="o">=</span> <span class="p">(</span><span class="n">seg_len</span><span class="p">[:,</span> <span class="kc">None</span><span class="p">]</span> <span class="o">+</span> <span class="n">seg_len</span><span class="p">[</span><span class="kc">None</span><span class="p">,</span> <span class="p">:])</span> <span class="o">/</span> <span class="mi">2</span>
  <span class="n">seg_mid_z</span> <span class="o">=</span> <span class="n">seg_mid</span><span class="p">[:,</span> <span class="mi">2</span><span class="p">]</span>
  <span class="n">seg_mid_dist</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">sqrt</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">square</span><span class="p">(</span><span class="n">seg_mid</span><span class="p">[:,</span> <span class="kc">None</span><span class="p">]</span> <span class="o">-</span> <span class="n">seg_mid</span><span class="p">[</span><span class="kc">None</span><span class="p">,</span> <span class="p">:])</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="o">-</span><span class="mi">1</span><span class="p">))</span>
  <span class="n">shadow_mask</span> <span class="o">=</span> <span class="n">sigmoid</span><span class="p">(</span><span class="n">seg_len_cutoff</span> <span class="o">*</span> <span class="mf">2.0</span> <span class="o">-</span> <span class="n">seg_mid_dist</span><span class="p">)</span> <span class="o">*</span> <span class="p">(</span><span class="n">seg_mid_z</span><span class="p">[:,</span> <span class="kc">None</span><span class="p">]</span> <span class="o">&lt;</span> <span class="n">seg_mid_z</span><span class="p">[</span><span class="kc">None</span><span class="p">,</span> <span class="p">:])</span>
  <span class="n">np</span><span class="o">.</span><span class="n">fill_diagonal</span><span class="p">(</span><span class="n">shadow_mask</span><span class="p">,</span> <span class="mf">0.0</span><span class="p">)</span>
  <span class="n">shadow_mask</span> <span class="o">=</span> <span class="n">shadow</span> <span class="o">**</span> <span class="n">shadow_mask</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="o">-</span><span class="mi">1</span><span class="p">,</span> <span class="n">keepdims</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

  <span class="n">seg_mid_xz</span> <span class="o">=</span> <span class="n">seg_mid</span><span class="p">[:,</span> <span class="p">:</span><span class="mi">2</span><span class="p">]</span>
  <span class="n">seg_mid_xydist</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">sqrt</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">square</span><span class="p">(</span><span class="n">seg_mid_xz</span><span class="p">[:,</span> <span class="kc">None</span><span class="p">]</span> <span class="o">-</span> <span class="n">seg_mid_xz</span><span class="p">[</span><span class="kc">None</span><span class="p">,</span> <span class="p">:])</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="o">-</span><span class="mi">1</span><span class="p">))</span>
  <span class="n">tint_mask</span> <span class="o">=</span> <span class="n">sigmoid</span><span class="p">(</span><span class="n">seg_len_cutoff</span> <span class="o">/</span> <span class="mi">2</span> <span class="o">-</span> <span class="n">seg_mid_xydist</span><span class="p">)</span> <span class="o">*</span> <span class="p">(</span><span class="n">seg_mid_z</span><span class="p">[:,</span> <span class="kc">None</span><span class="p">]</span> <span class="o">&lt;</span> <span class="n">seg_mid_z</span><span class="p">[</span><span class="kc">None</span><span class="p">,</span> <span class="p">:])</span>
  <span class="n">np</span><span class="o">.</span><span class="n">fill_diagonal</span><span class="p">(</span><span class="n">tint_mask</span><span class="p">,</span> <span class="mf">0.0</span><span class="p">)</span>
  <span class="n">tint_mask</span> <span class="o">=</span> <span class="mi">1</span> <span class="o">-</span> <span class="n">tint_mask</span><span class="o">.</span><span class="n">max</span><span class="p">(</span><span class="o">-</span><span class="mi">1</span><span class="p">,</span> <span class="n">keepdims</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

  <span class="n">colors</span><span class="p">[:,</span> <span class="p">:</span><span class="mi">3</span><span class="p">]</span> <span class="o">=</span> <span class="n">colors</span><span class="p">[:,</span> <span class="p">:</span><span class="mi">3</span><span class="p">]</span> <span class="o">+</span> <span class="p">(</span><span class="mi">1</span> <span class="o">-</span> <span class="n">colors</span><span class="p">[:,</span> <span class="p">:</span><span class="mi">3</span><span class="p">])</span> <span class="o">*</span> <span class="p">(</span><span class="mf">0.50</span> <span class="o">*</span> <span class="n">z</span> <span class="o">+</span> <span class="mf">0.50</span> <span class="o">*</span> <span class="n">tint_mask</span><span class="p">)</span> <span class="o">/</span> <span class="mi">3</span>
  <span class="n">colors</span><span class="p">[:,</span> <span class="p">:</span><span class="mi">3</span><span class="p">]</span> <span class="o">=</span> <span class="n">colors</span><span class="p">[:,</span> <span class="p">:</span><span class="mi">3</span><span class="p">]</span> <span class="o">*</span> <span class="p">(</span><span class="mf">0.20</span> <span class="o">+</span> <span class="mf">0.25</span> <span class="o">*</span> <span class="n">z</span> <span class="o">+</span> <span class="mf">0.55</span> <span class="o">*</span> <span class="n">shadow_mask</span><span class="p">)</span>

  <span class="n">set_lim</span> <span class="o">=</span> <span class="kc">False</span>
  <span class="k">if</span> <span class="n">ax</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
    <span class="n">fig</span><span class="p">,</span> <span class="n">ax</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">()</span>
    <span class="n">fig</span><span class="o">.</span><span class="n">set_figwidth</span><span class="p">(</span><span class="mi">5</span><span class="p">)</span>
    <span class="n">fig</span><span class="o">.</span><span class="n">set_figheight</span><span class="p">(</span><span class="mi">5</span><span class="p">)</span>
    <span class="n">set_lim</span> <span class="o">=</span> <span class="kc">True</span>
  <span class="k">else</span><span class="p">:</span>
    <span class="n">fig</span> <span class="o">=</span> <span class="n">ax</span><span class="o">.</span><span class="n">get_figure</span><span class="p">()</span>
    <span class="k">if</span> <span class="n">ax</span><span class="o">.</span><span class="n">get_xlim</span><span class="p">()</span> <span class="o">==</span> <span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">):</span>
      <span class="n">set_lim</span> <span class="o">=</span> <span class="kc">True</span>

  <span class="k">if</span> <span class="n">set_lim</span><span class="p">:</span>
    <span class="n">xy_min</span> <span class="o">=</span> <span class="n">xyz</span><span class="p">[:,</span> <span class="p">:</span><span class="mi">2</span><span class="p">]</span><span class="o">.</span><span class="n">min</span><span class="p">()</span> <span class="o">-</span> <span class="n">line_w</span>
    <span class="n">xy_max</span> <span class="o">=</span> <span class="n">xyz</span><span class="p">[:,</span> <span class="p">:</span><span class="mi">2</span><span class="p">]</span><span class="o">.</span><span class="n">max</span><span class="p">()</span> <span class="o">+</span> <span class="n">line_w</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">set_xlim</span><span class="p">(</span><span class="n">xy_min</span><span class="p">,</span> <span class="n">xy_max</span><span class="p">)</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">set_ylim</span><span class="p">(</span><span class="n">xy_min</span><span class="p">,</span> <span class="n">xy_max</span><span class="p">)</span>

  <span class="n">ax</span><span class="o">.</span><span class="n">set_aspect</span><span class="p">(</span><span class="s2">&quot;equal&quot;</span><span class="p">)</span>
  <span class="n">ax</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="s2">&quot;Distance ()&quot;</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="mi">12</span><span class="p">)</span>
  <span class="n">ax</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="s2">&quot;Distance ()&quot;</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="mi">12</span><span class="p">)</span>

  <span class="c1"># determine linewidths</span>
  <span class="n">width</span> <span class="o">=</span> <span class="n">fig</span><span class="o">.</span><span class="n">bbox_inches</span><span class="o">.</span><span class="n">width</span> <span class="o">*</span> <span class="n">ax</span><span class="o">.</span><span class="n">get_position</span><span class="p">()</span><span class="o">.</span><span class="n">width</span>
  <span class="n">linewidths</span> <span class="o">=</span> <span class="n">line_w</span> <span class="o">*</span> <span class="mi">72</span> <span class="o">*</span> <span class="n">width</span> <span class="o">/</span> <span class="n">np</span><span class="o">.</span><span class="n">diff</span><span class="p">(</span><span class="n">ax</span><span class="o">.</span><span class="n">get_xlim</span><span class="p">())</span>

  <span class="n">lines</span> <span class="o">=</span> <span class="n">mcoll</span><span class="o">.</span><span class="n">LineCollection</span><span class="p">(</span>
    <span class="n">seg_xy</span><span class="p">[</span><span class="n">order</span><span class="p">],</span> <span class="n">colors</span><span class="o">=</span><span class="n">colors</span><span class="p">[</span><span class="n">order</span><span class="p">],</span> <span class="n">linewidths</span><span class="o">=</span><span class="n">linewidths</span><span class="p">,</span> <span class="n">path_effects</span><span class="o">=</span><span class="p">[</span><span class="n">matplotlib</span><span class="o">.</span><span class="n">patheffects</span><span class="o">.</span><span class="n">Stroke</span><span class="p">(</span><span class="n">capstyle</span><span class="o">=</span><span class="s2">&quot;round&quot;</span><span class="p">)]</span>
  <span class="p">)</span>
  <span class="k">return</span> <span class="n">ax</span><span class="o">.</span><span class="n">add_collection</span><span class="p">(</span><span class="n">lines</span><span class="p">)</span></div>



<div class="viewcode-block" id="animate_pseudo_3D">
<a class="viewcode-back" href="../../neurosnap.html#neurosnap.protein.animate_pseudo_3D">[docs]</a>
<span class="k">def</span><span class="w"> </span><span class="nf">animate_pseudo_3D</span><span class="p">(</span>
  <span class="n">fig</span><span class="p">:</span> <span class="n">matplotlib</span><span class="o">.</span><span class="n">figure</span><span class="o">.</span><span class="n">Figure</span><span class="p">,</span>
  <span class="n">ax</span><span class="p">:</span> <span class="n">matplotlib</span><span class="o">.</span><span class="n">axes</span><span class="o">.</span><span class="n">Axes</span><span class="p">,</span>
  <span class="n">frames</span><span class="p">:</span> <span class="n">matplotlib</span><span class="o">.</span><span class="n">collections</span><span class="o">.</span><span class="n">LineCollection</span><span class="p">,</span>
  <span class="n">titles</span><span class="p">:</span> <span class="n">Union</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]]</span> <span class="o">=</span> <span class="s2">&quot;Protein Animation&quot;</span><span class="p">,</span>
  <span class="n">interval</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">200</span><span class="p">,</span>
  <span class="n">repeat_delay</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">0</span><span class="p">,</span>
  <span class="n">repeat</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">matplotlib</span><span class="o">.</span><span class="n">animation</span><span class="o">.</span><span class="n">ArtistAnimation</span><span class="p">:</span>
<span class="w">  </span><span class="sd">&quot;&quot;&quot;Animate multiple Pseudo 3D LineCollection objects.</span>

<span class="sd">  Parameters:</span>
<span class="sd">    fig: Matplotlib figure that contains all the frames</span>
<span class="sd">    ax: Matplotlib axes for the figure that contains all the frames</span>
<span class="sd">    frames: List of LineCollection objects</span>
<span class="sd">    titles: Single title or list of titles corresponding to each frame</span>
<span class="sd">    interval: Delay between frames in milliseconds</span>
<span class="sd">    repeat_delay: The delay in milliseconds between consecutive animation runs, if repeat is True</span>
<span class="sd">    repeat: Whether the animation repeats when the sequence of frames is completed</span>

<span class="sd">  Returns:</span>
<span class="sd">    Animation of all the different frames</span>

<span class="sd">  &quot;&quot;&quot;</span>
  <span class="c1"># check titles</span>
  <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">titles</span><span class="p">,</span> <span class="nb">str</span><span class="p">):</span>
    <span class="n">titles</span> <span class="o">=</span> <span class="p">[</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">titles</span><span class="si">}</span><span class="s2"> (</span><span class="si">{</span><span class="n">i</span><span class="o">+</span><span class="mi">1</span><span class="si">}</span><span class="s2">/</span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">frames</span><span class="p">)</span><span class="si">}</span><span class="s2">)&quot;</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">frames</span><span class="p">))]</span>
  <span class="k">elif</span> <span class="nb">len</span><span class="p">(</span><span class="n">titles</span><span class="p">)</span> <span class="o">==</span> <span class="nb">len</span><span class="p">(</span><span class="n">frames</span><span class="p">):</span>
    <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;The number of titles (</span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">titles</span><span class="p">)</span><span class="si">}</span><span class="s2">) does not match the number of frames (</span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">frames</span><span class="p">)</span><span class="si">}</span><span class="s2">).&quot;</span><span class="p">)</span>

  <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">titles</span><span class="p">,</span> <span class="nb">str</span><span class="p">):</span>
    <span class="n">titles</span> <span class="o">=</span> <span class="p">[</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">titles</span><span class="si">}</span><span class="s2"> (</span><span class="si">{</span><span class="n">i</span><span class="o">+</span><span class="mi">1</span><span class="si">}</span><span class="s2">/</span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">frames</span><span class="p">)</span><span class="si">}</span><span class="s2">)&quot;</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">frames</span><span class="p">))]</span>
  <span class="k">elif</span> <span class="nb">len</span><span class="p">(</span><span class="n">titles</span><span class="p">)</span> <span class="o">==</span> <span class="nb">len</span><span class="p">(</span><span class="n">frames</span><span class="p">):</span>
    <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;The number of titles (</span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">titles</span><span class="p">)</span><span class="si">}</span><span class="s2">) does not match the number of frames (</span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">frames</span><span class="p">)</span><span class="si">}</span><span class="s2">).&quot;</span><span class="p">)</span>

  <span class="c1"># Gather all vertices safely across multiple paths per frame</span>
  <span class="n">all_x</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">concatenate</span><span class="p">([</span><span class="n">path</span><span class="o">.</span><span class="n">vertices</span><span class="p">[:,</span> <span class="mi">0</span><span class="p">]</span> <span class="k">for</span> <span class="n">frame</span> <span class="ow">in</span> <span class="n">frames</span> <span class="k">for</span> <span class="n">path</span> <span class="ow">in</span> <span class="n">frame</span><span class="o">.</span><span class="n">get_paths</span><span class="p">()])</span>
  <span class="n">all_y</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">concatenate</span><span class="p">([</span><span class="n">path</span><span class="o">.</span><span class="n">vertices</span><span class="p">[:,</span> <span class="mi">1</span><span class="p">]</span> <span class="k">for</span> <span class="n">frame</span> <span class="ow">in</span> <span class="n">frames</span> <span class="k">for</span> <span class="n">path</span> <span class="ow">in</span> <span class="n">frame</span><span class="o">.</span><span class="n">get_paths</span><span class="p">()])</span>
  <span class="c1"># Calculate limits with optional padding</span>
  <span class="n">x_min</span><span class="p">,</span> <span class="n">x_max</span> <span class="o">=</span> <span class="n">all_x</span><span class="o">.</span><span class="n">min</span><span class="p">(),</span> <span class="n">all_x</span><span class="o">.</span><span class="n">max</span><span class="p">()</span>
  <span class="n">y_min</span><span class="p">,</span> <span class="n">y_max</span> <span class="o">=</span> <span class="n">all_y</span><span class="o">.</span><span class="n">min</span><span class="p">(),</span> <span class="n">all_y</span><span class="o">.</span><span class="n">max</span><span class="p">()</span>
  <span class="c1"># Apply padding to the limits</span>
  <span class="n">padding</span> <span class="o">=</span> <span class="mi">4</span>
  <span class="n">ax</span><span class="o">.</span><span class="n">set_xlim</span><span class="p">(</span><span class="n">x_min</span> <span class="o">-</span> <span class="n">padding</span><span class="p">,</span> <span class="n">x_max</span> <span class="o">+</span> <span class="n">padding</span><span class="p">)</span>
  <span class="n">ax</span><span class="o">.</span><span class="n">set_ylim</span><span class="p">(</span><span class="n">y_min</span> <span class="o">-</span> <span class="n">padding</span><span class="p">,</span> <span class="n">y_max</span> <span class="o">+</span> <span class="n">padding</span><span class="p">)</span>
  <span class="c1"># disable axes</span>
  <span class="n">ax</span><span class="o">.</span><span class="n">axis</span><span class="p">(</span><span class="s2">&quot;off&quot;</span><span class="p">)</span>

  <span class="k">def</span><span class="w"> </span><span class="nf">init</span><span class="p">():</span>
    <span class="c1"># hide all frames</span>
    <span class="k">for</span> <span class="n">frame</span> <span class="ow">in</span> <span class="n">frames</span><span class="p">:</span>
      <span class="n">frame</span><span class="o">.</span><span class="n">set_visible</span><span class="p">(</span><span class="kc">False</span><span class="p">)</span>
    <span class="c1"># Initialize the plot with the first frame</span>
    <span class="n">collection</span> <span class="o">=</span> <span class="n">frames</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">add_collection</span><span class="p">(</span><span class="n">collection</span><span class="p">)</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="n">titles</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
    <span class="k">return</span> <span class="p">(</span><span class="n">collection</span><span class="p">,)</span>

  <span class="k">def</span><span class="w"> </span><span class="nf">animate</span><span class="p">(</span><span class="n">i</span><span class="p">):</span>
    <span class="n">frames</span><span class="p">[</span><span class="n">i</span> <span class="o">-</span> <span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">set_visible</span><span class="p">(</span><span class="kc">False</span><span class="p">)</span>
    <span class="n">frames</span><span class="p">[</span><span class="n">i</span><span class="p">]</span><span class="o">.</span><span class="n">set_visible</span><span class="p">(</span><span class="kc">True</span><span class="p">)</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">add_collection</span><span class="p">(</span><span class="n">frames</span><span class="p">[</span><span class="n">i</span><span class="p">])</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="n">titles</span><span class="p">[</span><span class="n">i</span><span class="p">])</span>
    <span class="k">return</span> <span class="p">(</span><span class="n">frames</span><span class="p">[</span><span class="n">i</span><span class="p">],)</span>

  <span class="n">ani</span> <span class="o">=</span> <span class="n">animation</span><span class="o">.</span><span class="n">FuncAnimation</span><span class="p">(</span><span class="n">fig</span><span class="p">,</span> <span class="n">animate</span><span class="p">,</span> <span class="n">init_func</span><span class="o">=</span><span class="n">init</span><span class="p">,</span> <span class="n">frames</span><span class="o">=</span><span class="nb">len</span><span class="p">(</span><span class="n">frames</span><span class="p">),</span> <span class="n">interval</span><span class="o">=</span><span class="n">interval</span><span class="p">,</span> <span class="n">repeat_delay</span><span class="o">=</span><span class="n">repeat_delay</span><span class="p">,</span> <span class="n">repeat</span><span class="o">=</span><span class="n">repeat</span><span class="p">)</span>
  <span class="k">return</span> <span class="n">ani</span></div>

</pre></div>

           </div>
          </div>
          <footer>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2022-2025, Neurosnap Inc..</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>